// @cloudflare/vitest-pool-workers/dist/pool/index.mjs
import { createRequire } from "node:module";
import assert from "node:assert";
import crypto from "node:crypto";
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath, pathToFileURL } from "node:url";
import net from "node:net";
import os from "node:os";
import { Log, LogLevel, Miniflare, ModuleRuleTypeSchema, PLUGINS, Response, compileModuleRules, formatZodError, getNodeCompat, getRootPath, kCurrentWorker, kUnsafeEphemeralUniqueKey, maybeApply, mergeWorkerOptions, parseWithRootPath, structuredSerializableReducers, structuredSerializableRevivers, supportedCompatibilityDate, testRegExps } from "miniflare";
import { experimental_readRawConfig } from "wrangler";
import { build } from "esbuild";
import fs$1 from "node:fs/promises";
import posixPath from "node:path/posix";
import util from "node:util";
import * as cjsModuleLexer from "cjs-module-lexer";
import { z } from "zod";

//#region rolldown:runtime
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __commonJS = (cb, mod) => function() {
	return mod || (0, cb[__getOwnPropNames(cb)[0]])((mod = { exports: {} }).exports, mod), mod.exports;
};
var __copyProps = (to, from, except, desc) => {
	if (from && typeof from === "object" || typeof from === "function") for (var keys = __getOwnPropNames(from), i = 0, n = keys.length, key; i < n; i++) {
		key = keys[i];
		if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
			get: ((k) => from[k]).bind(null, key),
			enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
		});
	}
	return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
	value: mod,
	enumerable: true
}) : target, mod));

//#endregion
//#region ../../node_modules/.pnpm/devalue@5.6.3/node_modules/devalue/src/utils.js
var DevalueError = class extends Error {
	/**
	* @param {string} message
	* @param {string[]} keys
	* @param {any} [value] - The value that failed to be serialized
	* @param {any} [root] - The root value being serialized
	*/
	constructor(message, keys, value, root) {
		super(message);
		this.name = "DevalueError";
		this.path = keys.join("");
		this.value = value;
		this.root = root;
	}
};
/** @param {any} thing */
function is_primitive(thing) {
	return Object(thing) !== thing;
}
const object_proto_names = /* @__PURE__ */ Object.getOwnPropertyNames(Object.prototype).sort().join("\0");
/** @param {any} thing */
function is_plain_object(thing) {
	const proto = Object.getPrototypeOf(thing);
	return proto === Object.prototype || proto === null || Object.getPrototypeOf(proto) === null || Object.getOwnPropertyNames(proto).sort().join("\0") === object_proto_names;
}
/** @param {any} thing */
function get_type(thing) {
	return Object.prototype.toString.call(thing).slice(8, -1);
}
/** @param {string} char */
function get_escaped_char(char) {
	switch (char) {
		case "\"": return "\\\"";
		case "<": return "\\u003C";
		case "\\": return "\\\\";
		case "\n": return "\\n";
		case "\r": return "\\r";
		case "	": return "\\t";
		case "\b": return "\\b";
		case "\f": return "\\f";
		case "\u2028": return "\\u2028";
		case "\u2029": return "\\u2029";
		default: return char < " " ? `\\u${char.charCodeAt(0).toString(16).padStart(4, "0")}` : "";
	}
}
/** @param {string} str */
function stringify_string(str) {
	let result = "";
	let last_pos = 0;
	const len = str.length;
	for (let i = 0; i < len; i += 1) {
		const char = str[i];
		const replacement = get_escaped_char(char);
		if (replacement) {
			result += str.slice(last_pos, i) + replacement;
			last_pos = i + 1;
		}
	}
	return `"${last_pos === 0 ? str : result + str.slice(last_pos)}"`;
}
/** @param {Record<string | symbol, any>} object */
function enumerable_symbols(object) {
	return Object.getOwnPropertySymbols(object).filter((symbol) => Object.getOwnPropertyDescriptor(object, symbol).enumerable);
}
const is_identifier = /^[a-zA-Z_$][a-zA-Z_$0-9]*$/;
/** @param {string} key */
function stringify_key(key) {
	return is_identifier.test(key) ? "." + key : "[" + JSON.stringify(key) + "]";
}
/** @param {string} s */
function is_valid_array_index(s) {
	if (s.length === 0) return false;
	if (s.length > 1 && s.charCodeAt(0) === 48) return false;
	for (let i = 0; i < s.length; i++) {
		const c = s.charCodeAt(i);
		if (c < 48 || c > 57) return false;
	}
	const n = +s;
	if (n >= 2 ** 32 - 1) return false;
	if (n < 0) return false;
	return true;
}
/**
* Finds the populated indices of an array.
* @param {unknown[]} array
*/
function valid_array_indices(array) {
	const keys = Object.keys(array);
	for (var i = keys.length - 1; i >= 0; i--) if (is_valid_array_index(keys[i])) break;
	keys.length = i + 1;
	return keys;
}

//#endregion
//#region ../../node_modules/.pnpm/devalue@5.6.3/node_modules/devalue/src/base64.js
/**
* Base64 Encodes an arraybuffer
* @param {ArrayBuffer} arraybuffer
* @returns {string}
*/
function encode64(arraybuffer) {
	const dv = new DataView(arraybuffer);
	let binaryString = "";
	for (let i = 0; i < arraybuffer.byteLength; i++) binaryString += String.fromCharCode(dv.getUint8(i));
	return binaryToAscii(binaryString);
}
/**
* Decodes a base64 string into an arraybuffer
* @param {string} string
* @returns {ArrayBuffer}
*/
function decode64(string) {
	const binaryString = asciiToBinary(string);
	const arraybuffer = new ArrayBuffer(binaryString.length);
	const dv = new DataView(arraybuffer);
	for (let i = 0; i < arraybuffer.byteLength; i++) dv.setUint8(i, binaryString.charCodeAt(i));
	return arraybuffer;
}
const KEY_STRING = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
/**
* Substitute for atob since it's deprecated in node.
* Does not do any input validation.
*
* @see https://github.com/jsdom/abab/blob/master/lib/atob.js
*
* @param {string} data
* @returns {string}
*/
function asciiToBinary(data) {
	if (data.length % 4 === 0) data = data.replace(/==?$/, "");
	let output = "";
	let buffer = 0;
	let accumulatedBits = 0;
	for (let i = 0; i < data.length; i++) {
		buffer <<= 6;
		buffer |= KEY_STRING.indexOf(data[i]);
		accumulatedBits += 6;
		if (accumulatedBits === 24) {
			output += String.fromCharCode((buffer & 16711680) >> 16);
			output += String.fromCharCode((buffer & 65280) >> 8);
			output += String.fromCharCode(buffer & 255);
			buffer = accumulatedBits = 0;
		}
	}
	if (accumulatedBits === 12) {
		buffer >>= 4;
		output += String.fromCharCode(buffer);
	} else if (accumulatedBits === 18) {
		buffer >>= 2;
		output += String.fromCharCode((buffer & 65280) >> 8);
		output += String.fromCharCode(buffer & 255);
	}
	return output;
}
/**
* Substitute for btoa since it's deprecated in node.
* Does not do any input validation.
*
* @see https://github.com/jsdom/abab/blob/master/lib/btoa.js
*
* @param {string} str
* @returns {string}
*/
function binaryToAscii(str) {
	let out = "";
	for (let i = 0; i < str.length; i += 3) {
		/** @type {[number, number, number, number]} */
		const groupsOfSix = [
			void 0,
			void 0,
			void 0,
			void 0
		];
		groupsOfSix[0] = str.charCodeAt(i) >> 2;
		groupsOfSix[1] = (str.charCodeAt(i) & 3) << 4;
		if (str.length > i + 1) {
			groupsOfSix[1] |= str.charCodeAt(i + 1) >> 4;
			groupsOfSix[2] = (str.charCodeAt(i + 1) & 15) << 2;
		}
		if (str.length > i + 2) {
			groupsOfSix[2] |= str.charCodeAt(i + 2) >> 6;
			groupsOfSix[3] = str.charCodeAt(i + 2) & 63;
		}
		for (let j = 0; j < groupsOfSix.length; j++) if (typeof groupsOfSix[j] === "undefined") out += "=";
		else out += KEY_STRING[groupsOfSix[j]];
	}
	return out;
}

//#endregion
//#region ../../node_modules/.pnpm/devalue@5.6.3/node_modules/devalue/src/constants.js
const UNDEFINED = -1;
const HOLE = -2;
const NAN = -3;
const POSITIVE_INFINITY = -4;
const NEGATIVE_INFINITY = -5;
const NEGATIVE_ZERO = -6;
const SPARSE = -7;

//#endregion
//#region ../../node_modules/.pnpm/devalue@5.6.3/node_modules/devalue/src/parse.js
/**
* Revive a value serialized with `devalue.stringify`
* @param {string} serialized
* @param {Record<string, (value: any) => any>} [revivers]
*/
function parse(serialized, revivers) {
	return unflatten(JSON.parse(serialized), revivers);
}
/**
* Revive a value flattened with `devalue.stringify`
* @param {number | any[]} parsed
* @param {Record<string, (value: any) => any>} [revivers]
*/
function unflatten(parsed, revivers) {
	if (typeof parsed === "number") return hydrate(parsed, true);
	if (!Array.isArray(parsed) || parsed.length === 0) throw new Error("Invalid input");
	const values = parsed;
	const hydrated = Array(values.length);
	/**
	* A set of values currently being hydrated with custom revivers,
	* used to detect invalid cyclical dependencies
	* @type {Set<number> | null}
	*/
	let hydrating = null;
	/**
	* @param {number} index
	* @returns {any}
	*/
	function hydrate(index, standalone = false) {
		if (index === UNDEFINED) return void 0;
		if (index === NAN) return NaN;
		if (index === POSITIVE_INFINITY) return Infinity;
		if (index === NEGATIVE_INFINITY) return -Infinity;
		if (index === NEGATIVE_ZERO) return -0;
		if (standalone || typeof index !== "number") throw new Error(`Invalid input`);
		if (index in hydrated) return hydrated[index];
		const value = values[index];
		if (!value || typeof value !== "object") hydrated[index] = value;
		else if (Array.isArray(value)) if (typeof value[0] === "string") {
			const type = value[0];
			const reviver = revivers && Object.hasOwn(revivers, type) ? revivers[type] : void 0;
			if (reviver) {
				let i = value[1];
				if (typeof i !== "number") i = values.push(value[1]) - 1;
				hydrating ??= /* @__PURE__ */ new Set();
				if (hydrating.has(i)) throw new Error("Invalid circular reference");
				hydrating.add(i);
				hydrated[index] = reviver(hydrate(i));
				hydrating.delete(i);
				return hydrated[index];
			}
			switch (type) {
				case "Date":
					hydrated[index] = new Date(value[1]);
					break;
				case "Set":
					const set = /* @__PURE__ */ new Set();
					hydrated[index] = set;
					for (let i = 1; i < value.length; i += 1) set.add(hydrate(value[i]));
					break;
				case "Map":
					const map = /* @__PURE__ */ new Map();
					hydrated[index] = map;
					for (let i = 1; i < value.length; i += 2) map.set(hydrate(value[i]), hydrate(value[i + 1]));
					break;
				case "RegExp":
					hydrated[index] = new RegExp(value[1], value[2]);
					break;
				case "Object":
					hydrated[index] = Object(value[1]);
					break;
				case "BigInt":
					hydrated[index] = BigInt(value[1]);
					break;
				case "null":
					const obj = Object.create(null);
					hydrated[index] = obj;
					for (let i = 1; i < value.length; i += 2) obj[value[i]] = hydrate(value[i + 1]);
					break;
				case "Int8Array":
				case "Uint8Array":
				case "Uint8ClampedArray":
				case "Int16Array":
				case "Uint16Array":
				case "Int32Array":
				case "Uint32Array":
				case "Float32Array":
				case "Float64Array":
				case "BigInt64Array":
				case "BigUint64Array": {
					if (values[value[1]][0] !== "ArrayBuffer") throw new Error("Invalid data");
					const TypedArrayConstructor = globalThis[type];
					const typedArray = new TypedArrayConstructor(hydrate(value[1]));
					hydrated[index] = value[2] !== void 0 ? typedArray.subarray(value[2], value[3]) : typedArray;
					break;
				}
				case "ArrayBuffer": {
					const base64 = value[1];
					if (typeof base64 !== "string") throw new Error("Invalid ArrayBuffer encoding");
					hydrated[index] = decode64(base64);
					break;
				}
				case "Temporal.Duration":
				case "Temporal.Instant":
				case "Temporal.PlainDate":
				case "Temporal.PlainTime":
				case "Temporal.PlainDateTime":
				case "Temporal.PlainMonthDay":
				case "Temporal.PlainYearMonth":
				case "Temporal.ZonedDateTime": {
					const temporalName = type.slice(9);
					hydrated[index] = Temporal[temporalName].from(value[1]);
					break;
				}
				case "URL":
					hydrated[index] = new URL(value[1]);
					break;
				case "URLSearchParams":
					hydrated[index] = new URLSearchParams(value[1]);
					break;
				default: throw new Error(`Unknown type ${type}`);
			}
		} else if (value[0] === SPARSE) {
			const len = value[1];
			const array = new Array(len);
			hydrated[index] = array;
			for (let i = 2; i < value.length; i += 2) {
				const idx = value[i];
				array[idx] = hydrate(value[i + 1]);
			}
		} else {
			const array = new Array(value.length);
			hydrated[index] = array;
			for (let i = 0; i < value.length; i += 1) {
				const n = value[i];
				if (n === HOLE) continue;
				array[i] = hydrate(n);
			}
		}
		else {
			/** @type {Record<string, any>} */
			const object = {};
			hydrated[index] = object;
			for (const key of Object.keys(value)) {
				if (key === "__proto__") throw new Error("Cannot parse an object with a `__proto__` property");
				const n = value[key];
				object[key] = hydrate(n);
			}
		}
		return hydrated[index];
	}
	return hydrate(0);
}

//#endregion
//#region ../../node_modules/.pnpm/devalue@5.6.3/node_modules/devalue/src/stringify.js
/**
* Turn a value into a JSON string that can be parsed with `devalue.parse`
* @param {any} value
* @param {Record<string, (value: any) => any>} [reducers]
*/
function stringify(value, reducers) {
	/** @type {any[]} */
	const stringified = [];
	/** @type {Map<any, number>} */
	const indexes = /* @__PURE__ */ new Map();
	/** @type {Array<{ key: string, fn: (value: any) => any }>} */
	const custom = [];
	if (reducers) for (const key of Object.getOwnPropertyNames(reducers)) custom.push({
		key,
		fn: reducers[key]
	});
	/** @type {string[]} */
	const keys = [];
	let p = 0;
	/** @param {any} thing */
	function flatten(thing) {
		if (thing === void 0) return UNDEFINED;
		if (Number.isNaN(thing)) return NAN;
		if (thing === Infinity) return POSITIVE_INFINITY;
		if (thing === -Infinity) return NEGATIVE_INFINITY;
		if (thing === 0 && 1 / thing < 0) return NEGATIVE_ZERO;
		if (indexes.has(thing)) return indexes.get(thing);
		const index$1 = p++;
		indexes.set(thing, index$1);
		for (const { key, fn } of custom) {
			const value$1 = fn(thing);
			if (value$1) {
				stringified[index$1] = `["${key}",${flatten(value$1)}]`;
				return index$1;
			}
		}
		if (typeof thing === "function") throw new DevalueError(`Cannot stringify a function`, keys, thing, value);
		let str = "";
		if (is_primitive(thing)) str = stringify_primitive(thing);
		else {
			const type = get_type(thing);
			switch (type) {
				case "Number":
				case "String":
				case "Boolean":
					str = `["Object",${stringify_primitive(thing)}]`;
					break;
				case "BigInt":
					str = `["BigInt",${thing}]`;
					break;
				case "Date":
					str = `["Date","${!isNaN(thing.getDate()) ? thing.toISOString() : ""}"]`;
					break;
				case "URL":
					str = `["URL",${stringify_string(thing.toString())}]`;
					break;
				case "URLSearchParams":
					str = `["URLSearchParams",${stringify_string(thing.toString())}]`;
					break;
				case "RegExp":
					const { source, flags } = thing;
					str = flags ? `["RegExp",${stringify_string(source)},"${flags}"]` : `["RegExp",${stringify_string(source)}]`;
					break;
				case "Array": {
					let mostly_dense = false;
					str = "[";
					for (let i = 0; i < thing.length; i += 1) {
						if (i > 0) str += ",";
						if (Object.hasOwn(thing, i)) {
							keys.push(`[${i}]`);
							str += flatten(thing[i]);
							keys.pop();
						} else if (mostly_dense) str += HOLE;
						else {
							const populated_keys = valid_array_indices(thing);
							const population = populated_keys.length;
							const d = String(thing.length).length;
							if ((thing.length - population) * 3 > 4 + d + population * (d + 1)) {
								str = "[" + SPARSE + "," + thing.length;
								for (let j = 0; j < populated_keys.length; j++) {
									const key = populated_keys[j];
									keys.push(`[${key}]`);
									str += "," + key + "," + flatten(thing[key]);
									keys.pop();
								}
								break;
							} else {
								mostly_dense = true;
								str += HOLE;
							}
						}
					}
					str += "]";
					break;
				}
				case "Set":
					str = "[\"Set\"";
					for (const value$1 of thing) str += `,${flatten(value$1)}`;
					str += "]";
					break;
				case "Map":
					str = "[\"Map\"";
					for (const [key, value$1] of thing) {
						keys.push(`.get(${is_primitive(key) ? stringify_primitive(key) : "..."})`);
						str += `,${flatten(key)},${flatten(value$1)}`;
						keys.pop();
					}
					str += "]";
					break;
				case "Int8Array":
				case "Uint8Array":
				case "Uint8ClampedArray":
				case "Int16Array":
				case "Uint16Array":
				case "Int32Array":
				case "Uint32Array":
				case "Float32Array":
				case "Float64Array":
				case "BigInt64Array":
				case "BigUint64Array": {
					/** @type {import("./types.js").TypedArray} */
					const typedArray = thing;
					str = "[\"" + type + "\"," + flatten(typedArray.buffer);
					const a = thing.byteOffset;
					const b = a + thing.byteLength;
					if (a > 0 || b !== typedArray.buffer.byteLength) {
						const m = +/(\d+)/.exec(type)[1] / 8;
						str += `,${a / m},${b / m}`;
					}
					str += "]";
					break;
				}
				case "ArrayBuffer":
					str = `["ArrayBuffer","${encode64(thing)}"]`;
					break;
				case "Temporal.Duration":
				case "Temporal.Instant":
				case "Temporal.PlainDate":
				case "Temporal.PlainTime":
				case "Temporal.PlainDateTime":
				case "Temporal.PlainMonthDay":
				case "Temporal.PlainYearMonth":
				case "Temporal.ZonedDateTime":
					str = `["${type}",${stringify_string(thing.toString())}]`;
					break;
				default:
					if (!is_plain_object(thing)) throw new DevalueError(`Cannot stringify arbitrary non-POJOs`, keys, thing, value);
					if (enumerable_symbols(thing).length > 0) throw new DevalueError(`Cannot stringify POJOs with symbolic keys`, keys, thing, value);
					if (Object.getPrototypeOf(thing) === null) {
						str = "[\"null\"";
						for (const key of Object.keys(thing)) {
							if (key === "__proto__") throw new DevalueError(`Cannot stringify objects with __proto__ keys`, keys, thing, value);
							keys.push(stringify_key(key));
							str += `,${stringify_string(key)},${flatten(thing[key])}`;
							keys.pop();
						}
						str += "]";
					} else {
						str = "{";
						let started = false;
						for (const key of Object.keys(thing)) {
							if (key === "__proto__") throw new DevalueError(`Cannot stringify objects with __proto__ keys`, keys, thing, value);
							if (started) str += ",";
							started = true;
							keys.push(stringify_key(key));
							str += `${stringify_string(key)}:${flatten(thing[key])}`;
							keys.pop();
						}
						str += "}";
					}
			}
		}
		stringified[index$1] = str;
		return index$1;
	}
	const index = flatten(value);
	if (index < 0) return `${index}`;
	return `[${stringified.join(",")}]`;
}
/**
* @param {any} thing
* @returns {string}
*/
function stringify_primitive(thing) {
	const type = typeof thing;
	if (type === "string") return stringify_string(thing);
	if (thing instanceof String) return stringify_string(thing.toString());
	if (thing === void 0) return UNDEFINED.toString();
	if (thing === 0 && 1 / thing < 0) return NEGATIVE_ZERO.toString();
	if (type === "bigint") return `["BigInt","${thing}"]`;
	return String(thing);
}

//#endregion
//#region ../../node_modules/.pnpm/get-port@7.1.0/node_modules/get-port/index.js
var Locked = class extends Error {
	constructor(port) {
		super(`${port} is locked`);
	}
};
const lockedPorts = {
	old: /* @__PURE__ */ new Set(),
	young: /* @__PURE__ */ new Set()
};
const releaseOldLockedPortsIntervalMs = 1e3 * 15;
const minPort = 1024;
const maxPort = 65535;
let timeout;
const getLocalHosts = () => {
	const interfaces = os.networkInterfaces();
	const results = new Set([void 0, "0.0.0.0"]);
	for (const _interface of Object.values(interfaces)) for (const config of _interface) results.add(config.address);
	return results;
};
const checkAvailablePort = (options) => new Promise((resolve$1, reject) => {
	const server = net.createServer();
	server.unref();
	server.on("error", reject);
	server.listen(options, () => {
		const { port } = server.address();
		server.close(() => {
			resolve$1(port);
		});
	});
});
const getAvailablePort = async (options, hosts) => {
	if (options.host || options.port === 0) return checkAvailablePort(options);
	for (const host of hosts) try {
		await checkAvailablePort({
			port: options.port,
			host
		});
	} catch (error) {
		if (!["EADDRNOTAVAIL", "EINVAL"].includes(error.code)) throw error;
	}
	return options.port;
};
const portCheckSequence = function* (ports) {
	if (ports) yield* ports;
	yield 0;
};
async function getPorts(options) {
	let ports;
	let exclude = /* @__PURE__ */ new Set();
	if (options) {
		if (options.port) ports = typeof options.port === "number" ? [options.port] : options.port;
		if (options.exclude) {
			const excludeIterable = options.exclude;
			if (typeof excludeIterable[Symbol.iterator] !== "function") throw new TypeError("The `exclude` option must be an iterable.");
			for (const element of excludeIterable) {
				if (typeof element !== "number") throw new TypeError("Each item in the `exclude` option must be a number corresponding to the port you want excluded.");
				if (!Number.isSafeInteger(element)) throw new TypeError(`Number ${element} in the exclude option is not a safe integer and can't be used`);
			}
			exclude = new Set(excludeIterable);
		}
	}
	if (timeout === void 0) {
		timeout = setTimeout(() => {
			timeout = void 0;
			lockedPorts.old = lockedPorts.young;
			lockedPorts.young = /* @__PURE__ */ new Set();
		}, releaseOldLockedPortsIntervalMs);
		if (timeout.unref) timeout.unref();
	}
	const hosts = getLocalHosts();
	for (const port of portCheckSequence(ports)) try {
		if (exclude.has(port)) continue;
		let availablePort = await getAvailablePort({
			...options,
			port
		}, hosts);
		while (lockedPorts.old.has(availablePort) || lockedPorts.young.has(availablePort)) {
			if (port !== 0) throw new Locked(port);
			availablePort = await getAvailablePort({
				...options,
				port
			}, hosts);
		}
		lockedPorts.young.add(availablePort);
		return availablePort;
	} catch (error) {
		if (!["EADDRINUSE", "EACCES"].includes(error.code) && !(error instanceof Locked)) throw error;
	}
	throw new Error("No available ports found");
}
function portNumbers(from, to) {
	if (!Number.isInteger(from) || !Number.isInteger(to)) throw new TypeError("`from` and `to` must be integer numbers");
	if (from < minPort || from > maxPort) throw new RangeError(`'from' must be between ${minPort} and ${maxPort}`);
	if (to < minPort || to > maxPort) throw new RangeError(`'to' must be between ${minPort} and ${maxPort}`);
	if (from > to) throw new RangeError("`to` must be greater than or equal to `from`");
	const generator = function* (from$1, to$1) {
		for (let port = from$1; port <= to$1; port++) yield port;
	};
	return generator(from, to);
}

//#endregion
//#region ../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/internal/lrucache.js
var require_lrucache = /* @__PURE__ */ __commonJS({ "../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/internal/lrucache.js": ((exports, module) => {
	var LRUCache = class {
		constructor() {
			this.max = 1e3;
			this.map = /* @__PURE__ */ new Map();
		}
		get(key) {
			const value = this.map.get(key);
			if (value === void 0) return;
			else {
				this.map.delete(key);
				this.map.set(key, value);
				return value;
			}
		}
		delete(key) {
			return this.map.delete(key);
		}
		set(key, value) {
			if (!this.delete(key) && value !== void 0) {
				if (this.map.size >= this.max) {
					const firstKey = this.map.keys().next().value;
					this.delete(firstKey);
				}
				this.map.set(key, value);
			}
			return this;
		}
	};
	module.exports = LRUCache;
}) });

//#endregion
//#region ../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/internal/parse-options.js
var require_parse_options = /* @__PURE__ */ __commonJS({ "../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/internal/parse-options.js": ((exports, module) => {
	const looseOption = Object.freeze({ loose: true });
	const emptyOpts = Object.freeze({});
	const parseOptions$3 = (options) => {
		if (!options) return emptyOpts;
		if (typeof options !== "object") return looseOption;
		return options;
	};
	module.exports = parseOptions$3;
}) });

//#endregion
//#region ../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/internal/constants.js
var require_constants = /* @__PURE__ */ __commonJS({ "../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/internal/constants.js": ((exports, module) => {
	const SEMVER_SPEC_VERSION = "2.0.0";
	const MAX_LENGTH$2 = 256;
	const MAX_SAFE_INTEGER$1 = Number.MAX_SAFE_INTEGER || 9007199254740991;
	const MAX_SAFE_COMPONENT_LENGTH$1 = 16;
	const MAX_SAFE_BUILD_LENGTH$1 = MAX_LENGTH$2 - 6;
	const RELEASE_TYPES = [
		"major",
		"premajor",
		"minor",
		"preminor",
		"patch",
		"prepatch",
		"prerelease"
	];
	module.exports = {
		MAX_LENGTH: MAX_LENGTH$2,
		MAX_SAFE_COMPONENT_LENGTH: MAX_SAFE_COMPONENT_LENGTH$1,
		MAX_SAFE_BUILD_LENGTH: MAX_SAFE_BUILD_LENGTH$1,
		MAX_SAFE_INTEGER: MAX_SAFE_INTEGER$1,
		RELEASE_TYPES,
		SEMVER_SPEC_VERSION,
		FLAG_INCLUDE_PRERELEASE: 1,
		FLAG_LOOSE: 2
	};
}) });

//#endregion
//#region ../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/internal/debug.js
var require_debug = /* @__PURE__ */ __commonJS({ "../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/internal/debug.js": ((exports, module) => {
	const debug$4 = typeof process === "object" && process.env && process.env.NODE_DEBUG && /\bsemver\b/i.test(process.env.NODE_DEBUG) ? (...args) => console.error("SEMVER", ...args) : () => {};
	module.exports = debug$4;
}) });

//#endregion
//#region ../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/internal/re.js
var require_re = /* @__PURE__ */ __commonJS({ "../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/internal/re.js": ((exports, module) => {
	const { MAX_SAFE_COMPONENT_LENGTH, MAX_SAFE_BUILD_LENGTH, MAX_LENGTH: MAX_LENGTH$1 } = require_constants();
	const debug$3 = require_debug();
	exports = module.exports = {};
	const re$3 = exports.re = [];
	const safeRe = exports.safeRe = [];
	const src$1 = exports.src = [];
	const safeSrc = exports.safeSrc = [];
	const t$3 = exports.t = {};
	let R = 0;
	const LETTERDASHNUMBER = "[a-zA-Z0-9-]";
	const safeRegexReplacements = [
		["\\s", 1],
		["\\d", MAX_LENGTH$1],
		[LETTERDASHNUMBER, MAX_SAFE_BUILD_LENGTH]
	];
	const makeSafeRegex = (value) => {
		for (const [token, max] of safeRegexReplacements) value = value.split(`${token}*`).join(`${token}{0,${max}}`).split(`${token}+`).join(`${token}{1,${max}}`);
		return value;
	};
	const createToken = (name, value, isGlobal) => {
		const safe = makeSafeRegex(value);
		const index = R++;
		debug$3(name, index, value);
		t$3[name] = index;
		src$1[index] = value;
		safeSrc[index] = safe;
		re$3[index] = new RegExp(value, isGlobal ? "g" : void 0);
		safeRe[index] = new RegExp(safe, isGlobal ? "g" : void 0);
	};
	createToken("NUMERICIDENTIFIER", "0|[1-9]\\d*");
	createToken("NUMERICIDENTIFIERLOOSE", "\\d+");
	createToken("NONNUMERICIDENTIFIER", `\\d*[a-zA-Z-]${LETTERDASHNUMBER}*`);
	createToken("MAINVERSION", `(${src$1[t$3.NUMERICIDENTIFIER]})\\.(${src$1[t$3.NUMERICIDENTIFIER]})\\.(${src$1[t$3.NUMERICIDENTIFIER]})`);
	createToken("MAINVERSIONLOOSE", `(${src$1[t$3.NUMERICIDENTIFIERLOOSE]})\\.(${src$1[t$3.NUMERICIDENTIFIERLOOSE]})\\.(${src$1[t$3.NUMERICIDENTIFIERLOOSE]})`);
	createToken("PRERELEASEIDENTIFIER", `(?:${src$1[t$3.NUMERICIDENTIFIER]}|${src$1[t$3.NONNUMERICIDENTIFIER]})`);
	createToken("PRERELEASEIDENTIFIERLOOSE", `(?:${src$1[t$3.NUMERICIDENTIFIERLOOSE]}|${src$1[t$3.NONNUMERICIDENTIFIER]})`);
	createToken("PRERELEASE", `(?:-(${src$1[t$3.PRERELEASEIDENTIFIER]}(?:\\.${src$1[t$3.PRERELEASEIDENTIFIER]})*))`);
	createToken("PRERELEASELOOSE", `(?:-?(${src$1[t$3.PRERELEASEIDENTIFIERLOOSE]}(?:\\.${src$1[t$3.PRERELEASEIDENTIFIERLOOSE]})*))`);
	createToken("BUILDIDENTIFIER", `${LETTERDASHNUMBER}+`);
	createToken("BUILD", `(?:\\+(${src$1[t$3.BUILDIDENTIFIER]}(?:\\.${src$1[t$3.BUILDIDENTIFIER]})*))`);
	createToken("FULLPLAIN", `v?${src$1[t$3.MAINVERSION]}${src$1[t$3.PRERELEASE]}?${src$1[t$3.BUILD]}?`);
	createToken("FULL", `^${src$1[t$3.FULLPLAIN]}$`);
	createToken("LOOSEPLAIN", `[v=\\s]*${src$1[t$3.MAINVERSIONLOOSE]}${src$1[t$3.PRERELEASELOOSE]}?${src$1[t$3.BUILD]}?`);
	createToken("LOOSE", `^${src$1[t$3.LOOSEPLAIN]}$`);
	createToken("GTLT", "((?:<|>)?=?)");
	createToken("XRANGEIDENTIFIERLOOSE", `${src$1[t$3.NUMERICIDENTIFIERLOOSE]}|x|X|\\*`);
	createToken("XRANGEIDENTIFIER", `${src$1[t$3.NUMERICIDENTIFIER]}|x|X|\\*`);
	createToken("XRANGEPLAIN", `[v=\\s]*(${src$1[t$3.XRANGEIDENTIFIER]})(?:\\.(${src$1[t$3.XRANGEIDENTIFIER]})(?:\\.(${src$1[t$3.XRANGEIDENTIFIER]})(?:${src$1[t$3.PRERELEASE]})?${src$1[t$3.BUILD]}?)?)?`);
	createToken("XRANGEPLAINLOOSE", `[v=\\s]*(${src$1[t$3.XRANGEIDENTIFIERLOOSE]})(?:\\.(${src$1[t$3.XRANGEIDENTIFIERLOOSE]})(?:\\.(${src$1[t$3.XRANGEIDENTIFIERLOOSE]})(?:${src$1[t$3.PRERELEASELOOSE]})?${src$1[t$3.BUILD]}?)?)?`);
	createToken("XRANGE", `^${src$1[t$3.GTLT]}\\s*${src$1[t$3.XRANGEPLAIN]}$`);
	createToken("XRANGELOOSE", `^${src$1[t$3.GTLT]}\\s*${src$1[t$3.XRANGEPLAINLOOSE]}$`);
	createToken("COERCEPLAIN", `(^|[^\\d])(\\d{1,${MAX_SAFE_COMPONENT_LENGTH}})(?:\\.(\\d{1,${MAX_SAFE_COMPONENT_LENGTH}}))?(?:\\.(\\d{1,${MAX_SAFE_COMPONENT_LENGTH}}))?`);
	createToken("COERCE", `${src$1[t$3.COERCEPLAIN]}(?:$|[^\\d])`);
	createToken("COERCEFULL", src$1[t$3.COERCEPLAIN] + `(?:${src$1[t$3.PRERELEASE]})?(?:${src$1[t$3.BUILD]})?(?:$|[^\\d])`);
	createToken("COERCERTL", src$1[t$3.COERCE], true);
	createToken("COERCERTLFULL", src$1[t$3.COERCEFULL], true);
	createToken("LONETILDE", "(?:~>?)");
	createToken("TILDETRIM", `(\\s*)${src$1[t$3.LONETILDE]}\\s+`, true);
	exports.tildeTrimReplace = "$1~";
	createToken("TILDE", `^${src$1[t$3.LONETILDE]}${src$1[t$3.XRANGEPLAIN]}$`);
	createToken("TILDELOOSE", `^${src$1[t$3.LONETILDE]}${src$1[t$3.XRANGEPLAINLOOSE]}$`);
	createToken("LONECARET", "(?:\\^)");
	createToken("CARETTRIM", `(\\s*)${src$1[t$3.LONECARET]}\\s+`, true);
	exports.caretTrimReplace = "$1^";
	createToken("CARET", `^${src$1[t$3.LONECARET]}${src$1[t$3.XRANGEPLAIN]}$`);
	createToken("CARETLOOSE", `^${src$1[t$3.LONECARET]}${src$1[t$3.XRANGEPLAINLOOSE]}$`);
	createToken("COMPARATORLOOSE", `^${src$1[t$3.GTLT]}\\s*(${src$1[t$3.LOOSEPLAIN]})$|^$`);
	createToken("COMPARATOR", `^${src$1[t$3.GTLT]}\\s*(${src$1[t$3.FULLPLAIN]})$|^$`);
	createToken("COMPARATORTRIM", `(\\s*)${src$1[t$3.GTLT]}\\s*(${src$1[t$3.LOOSEPLAIN]}|${src$1[t$3.XRANGEPLAIN]})`, true);
	exports.comparatorTrimReplace = "$1$2$3";
	createToken("HYPHENRANGE", `^\\s*(${src$1[t$3.XRANGEPLAIN]})\\s+-\\s+(${src$1[t$3.XRANGEPLAIN]})\\s*$`);
	createToken("HYPHENRANGELOOSE", `^\\s*(${src$1[t$3.XRANGEPLAINLOOSE]})\\s+-\\s+(${src$1[t$3.XRANGEPLAINLOOSE]})\\s*$`);
	createToken("STAR", "(<|>)?=?\\s*\\*");
	createToken("GTE0", "^\\s*>=\\s*0\\.0\\.0\\s*$");
	createToken("GTE0PRE", "^\\s*>=\\s*0\\.0\\.0-0\\s*$");
}) });

//#endregion
//#region ../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/internal/identifiers.js
var require_identifiers = /* @__PURE__ */ __commonJS({ "../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/internal/identifiers.js": ((exports, module) => {
	const numeric = /^[0-9]+$/;
	const compareIdentifiers$1 = (a, b) => {
		const anum = numeric.test(a);
		const bnum = numeric.test(b);
		if (anum && bnum) {
			a = +a;
			b = +b;
		}
		return a === b ? 0 : anum && !bnum ? -1 : bnum && !anum ? 1 : a < b ? -1 : 1;
	};
	const rcompareIdentifiers = (a, b) => compareIdentifiers$1(b, a);
	module.exports = {
		compareIdentifiers: compareIdentifiers$1,
		rcompareIdentifiers
	};
}) });

//#endregion
//#region ../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/classes/semver.js
var require_semver = /* @__PURE__ */ __commonJS({ "../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/classes/semver.js": ((exports, module) => {
	const debug$2 = require_debug();
	const { MAX_LENGTH, MAX_SAFE_INTEGER } = require_constants();
	const { safeRe: re$2, safeSrc: src, t: t$2 } = require_re();
	const parseOptions$2 = require_parse_options();
	const { compareIdentifiers } = require_identifiers();
	var SemVer$3 = class SemVer$3 {
		constructor(version, options) {
			options = parseOptions$2(options);
			if (version instanceof SemVer$3) if (version.loose === !!options.loose && version.includePrerelease === !!options.includePrerelease) return version;
			else version = version.version;
			else if (typeof version !== "string") throw new TypeError(`Invalid version. Must be a string. Got type "${typeof version}".`);
			if (version.length > MAX_LENGTH) throw new TypeError(`version is longer than ${MAX_LENGTH} characters`);
			debug$2("SemVer", version, options);
			this.options = options;
			this.loose = !!options.loose;
			this.includePrerelease = !!options.includePrerelease;
			const m = version.trim().match(options.loose ? re$2[t$2.LOOSE] : re$2[t$2.FULL]);
			if (!m) throw new TypeError(`Invalid Version: ${version}`);
			this.raw = version;
			this.major = +m[1];
			this.minor = +m[2];
			this.patch = +m[3];
			if (this.major > MAX_SAFE_INTEGER || this.major < 0) throw new TypeError("Invalid major version");
			if (this.minor > MAX_SAFE_INTEGER || this.minor < 0) throw new TypeError("Invalid minor version");
			if (this.patch > MAX_SAFE_INTEGER || this.patch < 0) throw new TypeError("Invalid patch version");
			if (!m[4]) this.prerelease = [];
			else this.prerelease = m[4].split(".").map((id) => {
				if (/^[0-9]+$/.test(id)) {
					const num = +id;
					if (num >= 0 && num < MAX_SAFE_INTEGER) return num;
				}
				return id;
			});
			this.build = m[5] ? m[5].split(".") : [];
			this.format();
		}
		format() {
			this.version = `${this.major}.${this.minor}.${this.patch}`;
			if (this.prerelease.length) this.version += `-${this.prerelease.join(".")}`;
			return this.version;
		}
		toString() {
			return this.version;
		}
		compare(other) {
			debug$2("SemVer.compare", this.version, this.options, other);
			if (!(other instanceof SemVer$3)) {
				if (typeof other === "string" && other === this.version) return 0;
				other = new SemVer$3(other, this.options);
			}
			if (other.version === this.version) return 0;
			return this.compareMain(other) || this.comparePre(other);
		}
		compareMain(other) {
			if (!(other instanceof SemVer$3)) other = new SemVer$3(other, this.options);
			return compareIdentifiers(this.major, other.major) || compareIdentifiers(this.minor, other.minor) || compareIdentifiers(this.patch, other.patch);
		}
		comparePre(other) {
			if (!(other instanceof SemVer$3)) other = new SemVer$3(other, this.options);
			if (this.prerelease.length && !other.prerelease.length) return -1;
			else if (!this.prerelease.length && other.prerelease.length) return 1;
			else if (!this.prerelease.length && !other.prerelease.length) return 0;
			let i = 0;
			do {
				const a = this.prerelease[i];
				const b = other.prerelease[i];
				debug$2("prerelease compare", i, a, b);
				if (a === void 0 && b === void 0) return 0;
				else if (b === void 0) return 1;
				else if (a === void 0) return -1;
				else if (a === b) continue;
				else return compareIdentifiers(a, b);
			} while (++i);
		}
		compareBuild(other) {
			if (!(other instanceof SemVer$3)) other = new SemVer$3(other, this.options);
			let i = 0;
			do {
				const a = this.build[i];
				const b = other.build[i];
				debug$2("build compare", i, a, b);
				if (a === void 0 && b === void 0) return 0;
				else if (b === void 0) return 1;
				else if (a === void 0) return -1;
				else if (a === b) continue;
				else return compareIdentifiers(a, b);
			} while (++i);
		}
		inc(release, identifier, identifierBase) {
			if (release.startsWith("pre")) {
				if (!identifier && identifierBase === false) throw new Error("invalid increment argument: identifier is empty");
				if (identifier) {
					const r = /* @__PURE__ */ new RegExp(`^${this.options.loose ? src[t$2.PRERELEASELOOSE] : src[t$2.PRERELEASE]}$`);
					const match = `-${identifier}`.match(r);
					if (!match || match[1] !== identifier) throw new Error(`invalid identifier: ${identifier}`);
				}
			}
			switch (release) {
				case "premajor":
					this.prerelease.length = 0;
					this.patch = 0;
					this.minor = 0;
					this.major++;
					this.inc("pre", identifier, identifierBase);
					break;
				case "preminor":
					this.prerelease.length = 0;
					this.patch = 0;
					this.minor++;
					this.inc("pre", identifier, identifierBase);
					break;
				case "prepatch":
					this.prerelease.length = 0;
					this.inc("patch", identifier, identifierBase);
					this.inc("pre", identifier, identifierBase);
					break;
				case "prerelease":
					if (this.prerelease.length === 0) this.inc("patch", identifier, identifierBase);
					this.inc("pre", identifier, identifierBase);
					break;
				case "release":
					if (this.prerelease.length === 0) throw new Error(`version ${this.raw} is not a prerelease`);
					this.prerelease.length = 0;
					break;
				case "major":
					if (this.minor !== 0 || this.patch !== 0 || this.prerelease.length === 0) this.major++;
					this.minor = 0;
					this.patch = 0;
					this.prerelease = [];
					break;
				case "minor":
					if (this.patch !== 0 || this.prerelease.length === 0) this.minor++;
					this.patch = 0;
					this.prerelease = [];
					break;
				case "patch":
					if (this.prerelease.length === 0) this.patch++;
					this.prerelease = [];
					break;
				case "pre": {
					const base = Number(identifierBase) ? 1 : 0;
					if (this.prerelease.length === 0) this.prerelease = [base];
					else {
						let i = this.prerelease.length;
						while (--i >= 0) if (typeof this.prerelease[i] === "number") {
							this.prerelease[i]++;
							i = -2;
						}
						if (i === -1) {
							if (identifier === this.prerelease.join(".") && identifierBase === false) throw new Error("invalid increment argument: identifier already exists");
							this.prerelease.push(base);
						}
					}
					if (identifier) {
						let prerelease = [identifier, base];
						if (identifierBase === false) prerelease = [identifier];
						if (compareIdentifiers(this.prerelease[0], identifier) === 0) {
							if (isNaN(this.prerelease[1])) this.prerelease = prerelease;
						} else this.prerelease = prerelease;
					}
					break;
				}
				default: throw new Error(`invalid increment argument: ${release}`);
			}
			this.raw = this.format();
			if (this.build.length) this.raw += `+${this.build.join(".")}`;
			return this;
		}
	};
	module.exports = SemVer$3;
}) });

//#endregion
//#region ../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/functions/compare.js
var require_compare = /* @__PURE__ */ __commonJS({ "../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/functions/compare.js": ((exports, module) => {
	const SemVer$2 = require_semver();
	const compare$6 = (a, b, loose) => new SemVer$2(a, loose).compare(new SemVer$2(b, loose));
	module.exports = compare$6;
}) });

//#endregion
//#region ../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/functions/eq.js
var require_eq = /* @__PURE__ */ __commonJS({ "../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/functions/eq.js": ((exports, module) => {
	const compare$5 = require_compare();
	const eq$1 = (a, b, loose) => compare$5(a, b, loose) === 0;
	module.exports = eq$1;
}) });

//#endregion
//#region ../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/functions/neq.js
var require_neq = /* @__PURE__ */ __commonJS({ "../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/functions/neq.js": ((exports, module) => {
	const compare$4 = require_compare();
	const neq$1 = (a, b, loose) => compare$4(a, b, loose) !== 0;
	module.exports = neq$1;
}) });

//#endregion
//#region ../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/functions/gt.js
var require_gt = /* @__PURE__ */ __commonJS({ "../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/functions/gt.js": ((exports, module) => {
	const compare$3 = require_compare();
	const gt$1 = (a, b, loose) => compare$3(a, b, loose) > 0;
	module.exports = gt$1;
}) });

//#endregion
//#region ../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/functions/gte.js
var require_gte = /* @__PURE__ */ __commonJS({ "../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/functions/gte.js": ((exports, module) => {
	const compare$2 = require_compare();
	const gte$1 = (a, b, loose) => compare$2(a, b, loose) >= 0;
	module.exports = gte$1;
}) });

//#endregion
//#region ../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/functions/lt.js
var require_lt = /* @__PURE__ */ __commonJS({ "../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/functions/lt.js": ((exports, module) => {
	const compare$1 = require_compare();
	const lt$1 = (a, b, loose) => compare$1(a, b, loose) < 0;
	module.exports = lt$1;
}) });

//#endregion
//#region ../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/functions/lte.js
var require_lte = /* @__PURE__ */ __commonJS({ "../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/functions/lte.js": ((exports, module) => {
	const compare = require_compare();
	const lte$1 = (a, b, loose) => compare(a, b, loose) <= 0;
	module.exports = lte$1;
}) });

//#endregion
//#region ../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/functions/cmp.js
var require_cmp = /* @__PURE__ */ __commonJS({ "../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/functions/cmp.js": ((exports, module) => {
	const eq = require_eq();
	const neq = require_neq();
	const gt = require_gt();
	const gte = require_gte();
	const lt = require_lt();
	const lte = require_lte();
	const cmp$1 = (a, op, b, loose) => {
		switch (op) {
			case "===":
				if (typeof a === "object") a = a.version;
				if (typeof b === "object") b = b.version;
				return a === b;
			case "!==":
				if (typeof a === "object") a = a.version;
				if (typeof b === "object") b = b.version;
				return a !== b;
			case "":
			case "=":
			case "==": return eq(a, b, loose);
			case "!=": return neq(a, b, loose);
			case ">": return gt(a, b, loose);
			case ">=": return gte(a, b, loose);
			case "<": return lt(a, b, loose);
			case "<=": return lte(a, b, loose);
			default: throw new TypeError(`Invalid operator: ${op}`);
		}
	};
	module.exports = cmp$1;
}) });

//#endregion
//#region ../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/classes/comparator.js
var require_comparator = /* @__PURE__ */ __commonJS({ "../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/classes/comparator.js": ((exports, module) => {
	const ANY = Symbol("SemVer ANY");
	var Comparator$1 = class Comparator$1 {
		static get ANY() {
			return ANY;
		}
		constructor(comp, options) {
			options = parseOptions$1(options);
			if (comp instanceof Comparator$1) if (comp.loose === !!options.loose) return comp;
			else comp = comp.value;
			comp = comp.trim().split(/\s+/).join(" ");
			debug$1("comparator", comp, options);
			this.options = options;
			this.loose = !!options.loose;
			this.parse(comp);
			if (this.semver === ANY) this.value = "";
			else this.value = this.operator + this.semver.version;
			debug$1("comp", this);
		}
		parse(comp) {
			const r = this.options.loose ? re$1[t$1.COMPARATORLOOSE] : re$1[t$1.COMPARATOR];
			const m = comp.match(r);
			if (!m) throw new TypeError(`Invalid comparator: ${comp}`);
			this.operator = m[1] !== void 0 ? m[1] : "";
			if (this.operator === "=") this.operator = "";
			if (!m[2]) this.semver = ANY;
			else this.semver = new SemVer$1(m[2], this.options.loose);
		}
		toString() {
			return this.value;
		}
		test(version) {
			debug$1("Comparator.test", version, this.options.loose);
			if (this.semver === ANY || version === ANY) return true;
			if (typeof version === "string") try {
				version = new SemVer$1(version, this.options);
			} catch (er) {
				return false;
			}
			return cmp(version, this.operator, this.semver, this.options);
		}
		intersects(comp, options) {
			if (!(comp instanceof Comparator$1)) throw new TypeError("a Comparator is required");
			if (this.operator === "") {
				if (this.value === "") return true;
				return new Range$2(comp.value, options).test(this.value);
			} else if (comp.operator === "") {
				if (comp.value === "") return true;
				return new Range$2(this.value, options).test(comp.semver);
			}
			options = parseOptions$1(options);
			if (options.includePrerelease && (this.value === "<0.0.0-0" || comp.value === "<0.0.0-0")) return false;
			if (!options.includePrerelease && (this.value.startsWith("<0.0.0") || comp.value.startsWith("<0.0.0"))) return false;
			if (this.operator.startsWith(">") && comp.operator.startsWith(">")) return true;
			if (this.operator.startsWith("<") && comp.operator.startsWith("<")) return true;
			if (this.semver.version === comp.semver.version && this.operator.includes("=") && comp.operator.includes("=")) return true;
			if (cmp(this.semver, "<", comp.semver, options) && this.operator.startsWith(">") && comp.operator.startsWith("<")) return true;
			if (cmp(this.semver, ">", comp.semver, options) && this.operator.startsWith("<") && comp.operator.startsWith(">")) return true;
			return false;
		}
	};
	module.exports = Comparator$1;
	const parseOptions$1 = require_parse_options();
	const { safeRe: re$1, t: t$1 } = require_re();
	const cmp = require_cmp();
	const debug$1 = require_debug();
	const SemVer$1 = require_semver();
	const Range$2 = require_range();
}) });

//#endregion
//#region ../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/classes/range.js
var require_range = /* @__PURE__ */ __commonJS({ "../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/classes/range.js": ((exports, module) => {
	const SPACE_CHARACTERS = /\s+/g;
	var Range$1 = class Range$1 {
		constructor(range, options) {
			options = parseOptions(options);
			if (range instanceof Range$1) if (range.loose === !!options.loose && range.includePrerelease === !!options.includePrerelease) return range;
			else return new Range$1(range.raw, options);
			if (range instanceof Comparator) {
				this.raw = range.value;
				this.set = [[range]];
				this.formatted = void 0;
				return this;
			}
			this.options = options;
			this.loose = !!options.loose;
			this.includePrerelease = !!options.includePrerelease;
			this.raw = range.trim().replace(SPACE_CHARACTERS, " ");
			this.set = this.raw.split("||").map((r) => this.parseRange(r.trim())).filter((c) => c.length);
			if (!this.set.length) throw new TypeError(`Invalid SemVer Range: ${this.raw}`);
			if (this.set.length > 1) {
				const first = this.set[0];
				this.set = this.set.filter((c) => !isNullSet(c[0]));
				if (this.set.length === 0) this.set = [first];
				else if (this.set.length > 1) {
					for (const c of this.set) if (c.length === 1 && isAny(c[0])) {
						this.set = [c];
						break;
					}
				}
			}
			this.formatted = void 0;
		}
		get range() {
			if (this.formatted === void 0) {
				this.formatted = "";
				for (let i = 0; i < this.set.length; i++) {
					if (i > 0) this.formatted += "||";
					const comps = this.set[i];
					for (let k = 0; k < comps.length; k++) {
						if (k > 0) this.formatted += " ";
						this.formatted += comps[k].toString().trim();
					}
				}
			}
			return this.formatted;
		}
		format() {
			return this.range;
		}
		toString() {
			return this.range;
		}
		parseRange(range) {
			const memoKey = ((this.options.includePrerelease && FLAG_INCLUDE_PRERELEASE) | (this.options.loose && FLAG_LOOSE)) + ":" + range;
			const cached = cache.get(memoKey);
			if (cached) return cached;
			const loose = this.options.loose;
			const hr = loose ? re[t.HYPHENRANGELOOSE] : re[t.HYPHENRANGE];
			range = range.replace(hr, hyphenReplace(this.options.includePrerelease));
			debug("hyphen replace", range);
			range = range.replace(re[t.COMPARATORTRIM], comparatorTrimReplace);
			debug("comparator trim", range);
			range = range.replace(re[t.TILDETRIM], tildeTrimReplace);
			debug("tilde trim", range);
			range = range.replace(re[t.CARETTRIM], caretTrimReplace);
			debug("caret trim", range);
			let rangeList = range.split(" ").map((comp) => parseComparator(comp, this.options)).join(" ").split(/\s+/).map((comp) => replaceGTE0(comp, this.options));
			if (loose) rangeList = rangeList.filter((comp) => {
				debug("loose invalid filter", comp, this.options);
				return !!comp.match(re[t.COMPARATORLOOSE]);
			});
			debug("range list", rangeList);
			const rangeMap = /* @__PURE__ */ new Map();
			const comparators = rangeList.map((comp) => new Comparator(comp, this.options));
			for (const comp of comparators) {
				if (isNullSet(comp)) return [comp];
				rangeMap.set(comp.value, comp);
			}
			if (rangeMap.size > 1 && rangeMap.has("")) rangeMap.delete("");
			const result = [...rangeMap.values()];
			cache.set(memoKey, result);
			return result;
		}
		intersects(range, options) {
			if (!(range instanceof Range$1)) throw new TypeError("a Range is required");
			return this.set.some((thisComparators) => {
				return isSatisfiable(thisComparators, options) && range.set.some((rangeComparators) => {
					return isSatisfiable(rangeComparators, options) && thisComparators.every((thisComparator) => {
						return rangeComparators.every((rangeComparator) => {
							return thisComparator.intersects(rangeComparator, options);
						});
					});
				});
			});
		}
		test(version) {
			if (!version) return false;
			if (typeof version === "string") try {
				version = new SemVer(version, this.options);
			} catch (er) {
				return false;
			}
			for (let i = 0; i < this.set.length; i++) if (testSet(this.set[i], version, this.options)) return true;
			return false;
		}
	};
	module.exports = Range$1;
	const cache = new (require_lrucache())();
	const parseOptions = require_parse_options();
	const Comparator = require_comparator();
	const debug = require_debug();
	const SemVer = require_semver();
	const { safeRe: re, t, comparatorTrimReplace, tildeTrimReplace, caretTrimReplace } = require_re();
	const { FLAG_INCLUDE_PRERELEASE, FLAG_LOOSE } = require_constants();
	const isNullSet = (c) => c.value === "<0.0.0-0";
	const isAny = (c) => c.value === "";
	const isSatisfiable = (comparators, options) => {
		let result = true;
		const remainingComparators = comparators.slice();
		let testComparator = remainingComparators.pop();
		while (result && remainingComparators.length) {
			result = remainingComparators.every((otherComparator) => {
				return testComparator.intersects(otherComparator, options);
			});
			testComparator = remainingComparators.pop();
		}
		return result;
	};
	const parseComparator = (comp, options) => {
		debug("comp", comp, options);
		comp = replaceCarets(comp, options);
		debug("caret", comp);
		comp = replaceTildes(comp, options);
		debug("tildes", comp);
		comp = replaceXRanges(comp, options);
		debug("xrange", comp);
		comp = replaceStars(comp, options);
		debug("stars", comp);
		return comp;
	};
	const isX = (id) => !id || id.toLowerCase() === "x" || id === "*";
	const replaceTildes = (comp, options) => {
		return comp.trim().split(/\s+/).map((c) => replaceTilde(c, options)).join(" ");
	};
	const replaceTilde = (comp, options) => {
		const r = options.loose ? re[t.TILDELOOSE] : re[t.TILDE];
		return comp.replace(r, (_, M, m, p, pr) => {
			debug("tilde", comp, _, M, m, p, pr);
			let ret;
			if (isX(M)) ret = "";
			else if (isX(m)) ret = `>=${M}.0.0 <${+M + 1}.0.0-0`;
			else if (isX(p)) ret = `>=${M}.${m}.0 <${M}.${+m + 1}.0-0`;
			else if (pr) {
				debug("replaceTilde pr", pr);
				ret = `>=${M}.${m}.${p}-${pr} <${M}.${+m + 1}.0-0`;
			} else ret = `>=${M}.${m}.${p} <${M}.${+m + 1}.0-0`;
			debug("tilde return", ret);
			return ret;
		});
	};
	const replaceCarets = (comp, options) => {
		return comp.trim().split(/\s+/).map((c) => replaceCaret(c, options)).join(" ");
	};
	const replaceCaret = (comp, options) => {
		debug("caret", comp, options);
		const r = options.loose ? re[t.CARETLOOSE] : re[t.CARET];
		const z$1 = options.includePrerelease ? "-0" : "";
		return comp.replace(r, (_, M, m, p, pr) => {
			debug("caret", comp, _, M, m, p, pr);
			let ret;
			if (isX(M)) ret = "";
			else if (isX(m)) ret = `>=${M}.0.0${z$1} <${+M + 1}.0.0-0`;
			else if (isX(p)) if (M === "0") ret = `>=${M}.${m}.0${z$1} <${M}.${+m + 1}.0-0`;
			else ret = `>=${M}.${m}.0${z$1} <${+M + 1}.0.0-0`;
			else if (pr) {
				debug("replaceCaret pr", pr);
				if (M === "0") if (m === "0") ret = `>=${M}.${m}.${p}-${pr} <${M}.${m}.${+p + 1}-0`;
				else ret = `>=${M}.${m}.${p}-${pr} <${M}.${+m + 1}.0-0`;
				else ret = `>=${M}.${m}.${p}-${pr} <${+M + 1}.0.0-0`;
			} else {
				debug("no pr");
				if (M === "0") if (m === "0") ret = `>=${M}.${m}.${p}${z$1} <${M}.${m}.${+p + 1}-0`;
				else ret = `>=${M}.${m}.${p}${z$1} <${M}.${+m + 1}.0-0`;
				else ret = `>=${M}.${m}.${p} <${+M + 1}.0.0-0`;
			}
			debug("caret return", ret);
			return ret;
		});
	};
	const replaceXRanges = (comp, options) => {
		debug("replaceXRanges", comp, options);
		return comp.split(/\s+/).map((c) => replaceXRange(c, options)).join(" ");
	};
	const replaceXRange = (comp, options) => {
		comp = comp.trim();
		const r = options.loose ? re[t.XRANGELOOSE] : re[t.XRANGE];
		return comp.replace(r, (ret, gtlt, M, m, p, pr) => {
			debug("xRange", comp, ret, gtlt, M, m, p, pr);
			const xM = isX(M);
			const xm = xM || isX(m);
			const xp = xm || isX(p);
			const anyX = xp;
			if (gtlt === "=" && anyX) gtlt = "";
			pr = options.includePrerelease ? "-0" : "";
			if (xM) if (gtlt === ">" || gtlt === "<") ret = "<0.0.0-0";
			else ret = "*";
			else if (gtlt && anyX) {
				if (xm) m = 0;
				p = 0;
				if (gtlt === ">") {
					gtlt = ">=";
					if (xm) {
						M = +M + 1;
						m = 0;
						p = 0;
					} else {
						m = +m + 1;
						p = 0;
					}
				} else if (gtlt === "<=") {
					gtlt = "<";
					if (xm) M = +M + 1;
					else m = +m + 1;
				}
				if (gtlt === "<") pr = "-0";
				ret = `${gtlt + M}.${m}.${p}${pr}`;
			} else if (xm) ret = `>=${M}.0.0${pr} <${+M + 1}.0.0-0`;
			else if (xp) ret = `>=${M}.${m}.0${pr} <${M}.${+m + 1}.0-0`;
			debug("xRange return", ret);
			return ret;
		});
	};
	const replaceStars = (comp, options) => {
		debug("replaceStars", comp, options);
		return comp.trim().replace(re[t.STAR], "");
	};
	const replaceGTE0 = (comp, options) => {
		debug("replaceGTE0", comp, options);
		return comp.trim().replace(re[options.includePrerelease ? t.GTE0PRE : t.GTE0], "");
	};
	const hyphenReplace = (incPr) => ($0, from, fM, fm, fp, fpr, fb, to, tM, tm, tp, tpr) => {
		if (isX(fM)) from = "";
		else if (isX(fm)) from = `>=${fM}.0.0${incPr ? "-0" : ""}`;
		else if (isX(fp)) from = `>=${fM}.${fm}.0${incPr ? "-0" : ""}`;
		else if (fpr) from = `>=${from}`;
		else from = `>=${from}${incPr ? "-0" : ""}`;
		if (isX(tM)) to = "";
		else if (isX(tm)) to = `<${+tM + 1}.0.0-0`;
		else if (isX(tp)) to = `<${tM}.${+tm + 1}.0-0`;
		else if (tpr) to = `<=${tM}.${tm}.${tp}-${tpr}`;
		else if (incPr) to = `<${tM}.${tm}.${+tp + 1}-0`;
		else to = `<=${to}`;
		return `${from} ${to}`.trim();
	};
	const testSet = (set, version, options) => {
		for (let i = 0; i < set.length; i++) if (!set[i].test(version)) return false;
		if (version.prerelease.length && !options.includePrerelease) {
			for (let i = 0; i < set.length; i++) {
				debug(set[i].semver);
				if (set[i].semver === Comparator.ANY) continue;
				if (set[i].semver.prerelease.length > 0) {
					const allowed = set[i].semver;
					if (allowed.major === version.major && allowed.minor === version.minor && allowed.patch === version.patch) return true;
				}
			}
			return false;
		}
		return true;
	};
}) });

//#endregion
//#region ../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/functions/satisfies.js
var require_satisfies = /* @__PURE__ */ __commonJS({ "../../node_modules/.pnpm/semver@7.7.1/node_modules/semver/functions/satisfies.js": ((exports, module) => {
	const Range = require_range();
	const satisfies = (version, range, options) => {
		try {
			range = new Range(range, options);
		} catch (er) {
			return false;
		}
		return range.test(version);
	};
	module.exports = satisfies;
}) });

//#endregion
//#region src/pool/compatibility-flag-assertions.ts
var import_satisfies = /* @__PURE__ */ __toESM(require_satisfies(), 1);
/**
* The `CompatibilityFlagAssertions` class provides methods to validate compatibility flags and dates
* within a project's configuration. It ensures that specific flags are either present
* or absent and that compatibility dates meet the required criteria.
*/
var CompatibilityFlagAssertions = class {
	#compatibilityDate;
	#compatibilityFlags;
	#optionsPath;
	#relativeProjectPath;
	#relativeWranglerConfigPath;
	constructor(options) {
		this.#compatibilityDate = options.compatibilityDate;
		this.#compatibilityFlags = options.compatibilityFlags;
		this.#optionsPath = options.optionsPath;
		this.#relativeProjectPath = options.relativeProjectPath;
		this.#relativeWranglerConfigPath = options.relativeWranglerConfigPath;
	}
	/**
	* Checks if a specific flag is present in the compatibilityFlags array.
	*/
	#flagExists(flag) {
		return this.#compatibilityFlags.includes(flag);
	}
	/**
	* Constructs the base of the error message.
	*
	* @example
	* In project /path/to/project
	*
	* @example
	* In project /path/to/project's configuration file wrangler.toml
	*/
	#buildErrorMessageBase() {
		let message = `In project ${this.#relativeProjectPath}`;
		if (this.#relativeWranglerConfigPath) message += `'s configuration file ${this.#relativeWranglerConfigPath}`;
		return message;
	}
	/**
	* Constructs the configuration path part of the error message.
	*/
	#buildConfigPath(setting) {
		if (this.#relativeWranglerConfigPath) return `\`${setting}\``;
		const camelCaseSetting = setting.replace(/_(\w)/g, (_, letter) => letter.toUpperCase());
		return `\`${this.#optionsPath}.${camelCaseSetting}\``;
	}
	isEnabled(enableFlag, disableFlag, defaultOnDate) {
		if (this.#flagExists(disableFlag)) return false;
		return this.#flagExists(enableFlag) || isDateSufficient(this.#compatibilityDate, defaultOnDate);
	}
	/**
	* Ensures that a specific enable flag is present or that the compatibility date meets the required date.
	*/
	assertIsEnabled({ enableFlag, disableFlag, defaultOnDate }) {
		if (this.#flagExists(disableFlag)) return {
			isValid: false,
			errorMessage: `${this.#buildErrorMessageBase()}, ${this.#buildConfigPath("compatibility_flags")} must not contain "${disableFlag}".\nThis flag is incompatible with \`@cloudflare/vitest-pool-workers\`.`
		};
		const enableFlagPresent = this.#flagExists(enableFlag);
		const dateSufficient = isDateSufficient(this.#compatibilityDate, defaultOnDate);
		if (!enableFlagPresent && !dateSufficient) {
			let errorMessage = `${this.#buildErrorMessageBase()}, ${this.#buildConfigPath("compatibility_flags")} must contain "${enableFlag}"`;
			if (defaultOnDate) errorMessage += `, or ${this.#buildConfigPath("compatibility_date")} must be >= "${defaultOnDate}".`;
			errorMessage += `\nThis flag is required to use \`@cloudflare/vitest-pool-workers\`.`;
			return {
				isValid: false,
				errorMessage
			};
		}
		return { isValid: true };
	}
	/**
	* Ensures that a any one of a given set of flags is present in the compatibility_flags array.
	*/
	assertAtLeastOneFlagExists(flags) {
		if (flags.length === 0 || flags.some((flag) => this.#flagExists(flag))) return { isValid: true };
		return {
			isValid: false,
			errorMessage: `${this.#buildErrorMessageBase()}, ${this.#buildConfigPath("compatibility_flags")} must contain one of ${flags.map((flag) => `"${flag}"`).join("/")}.\nEither one of these flags is required to use \`@cloudflare/vitest-pool-workers\`.`
		};
	}
};
/**
* Parses a date string into a Date object.
*/
function parseDate(dateStr) {
	const date = new Date(dateStr);
	if (isNaN(date.getTime())) throw new Error(`Invalid date format: "${dateStr}"`);
	return date;
}
/**
* Checks if the compatibility date meets or exceeds the required date.
*/
function isDateSufficient(compatibilityDate, defaultOnDate) {
	if (!compatibilityDate || !defaultOnDate) return false;
	return parseDate(compatibilityDate) >= parseDate(defaultOnDate);
}

//#endregion
//#region src/pool/guess-exports.ts
/**
* Guesses the exports of the main Worker.
*
* We do this by running a lightweight esbuild of the main script, and looking at the meta-file generated by esbuild.
*
* - This doesn't support Python Workers, but then neither does Vitest.
* - Workers that need to be built with Vite plugins (e.g. containing wildcard re-exported virtual imports)
*   may not work with esbuild and so we won't be able to find the exports. But in these cases, the user can
*   always manually specify the exports in the Wrangler config via Service bindings, Workflows, or Durable Object migrations.
*
* @param entryFile The entry file of the Worker.
* @param additionalExports Additional exports to include in the export map.
*
* @returns A map of export names to their types (WorkerEntrypoint, DurableObject, WorkflowEntrypoint), or null if unknown.
*/
async function guessWorkerExports(entryFile, additionalExports = {}) {
	const exportMap = new Map(Object.entries(additionalExports));
	if (entryFile === void 0) return exportMap;
	if (path.extname(entryFile) == ".py") return exportMap;
	const result = await build({
		target: "es2024",
		loader: {
			".js": "jsx",
			".mjs": "jsx",
			".cjs": "jsx"
		},
		entryPoints: [path.resolve(entryFile)],
		metafile: true,
		bundle: false,
		write: false,
		logLevel: "silent"
	});
	const entryPoints = Object.values(result.metafile.outputs).filter((output) => output.entryPoint !== void 0);
	if (entryPoints.length === 0) throw new Error(`Cannot find entry-point "${entryFile}" in generated bundle.`);
	if (entryPoints.length > 1) throw new Error("More than one entry-point found for generated bundle.\n" + entryPoints.map((output) => ` - ${output.entryPoint}`).join("\n"));
	for (const exportName of entryPoints[0].exports) if (!exportMap.has(exportName)) exportMap.set(exportName, null);
	return exportMap;
}

//#endregion
//#region src/pool/helpers.ts
const WORKER_NAME_PREFIX = "vitest-pool-workers-";
function isFileNotFoundError(e) {
	return typeof e === "object" && e !== null && "code" in e && e.code === "ENOENT";
}
function getProjectPath(project) {
	return project.config.root;
}
function getRelativeProjectPath(project) {
	const projectPath = getProjectPath(project);
	return path.relative("", projectPath);
}
function getRelativeProjectConfigPath(project) {
	return project.config.config ? path.relative("", project.config.config) : getRelativeProjectPath(project);
}

//#endregion
//#region src/pool/loopback.ts
async function handleSnapshotRequest(request, url) {
	const filePath = url.searchParams.get("path");
	if (filePath === null) return new Response(null, { status: 400 });
	if (request.method === "POST") {
		await fs$1.mkdir(filePath, { recursive: true });
		return new Response(null, { status: 204 });
	}
	if (request.method === "PUT") {
		const snapshot = await request.arrayBuffer();
		await fs$1.mkdir(path.dirname(filePath), { recursive: true });
		await fs$1.writeFile(filePath, new Uint8Array(snapshot));
		return new Response(null, { status: 204 });
	}
	if (request.method === "GET") try {
		return new Response(await fs$1.readFile(filePath));
	} catch (e) {
		if (!isFileNotFoundError(e)) throw e;
		return new Response(null, { status: 404 });
	}
	if (request.method === "DELETE") {
		try {
			await fs$1.unlink(filePath);
		} catch (e) {
			if (!isFileNotFoundError(e)) throw e;
		}
		return new Response(null, { status: 204 });
	}
	return new Response(null, { status: 405 });
}
async function listDurableObjectIds(request, mf, url) {
	if (request.method !== "GET") return new Response(null, { status: 405 });
	const durableObjectPersistPath = mf.unsafeGetPersistPaths().get("do");
	assert(durableObjectPersistPath !== void 0, "Expected Durable Object persist path");
	const uniqueKey = url.searchParams.get("unique_key");
	if (uniqueKey === null) return new Response(null, { status: 400 });
	const namespacePath = path.join(durableObjectPersistPath, uniqueKey);
	const ids = [];
	try {
		const names = await fs$1.readdir(namespacePath);
		for (const name of names) if (name.endsWith(".sqlite")) ids.push(name.substring(0, name.length - 7));
	} catch (e) {
		if (!isFileNotFoundError(e)) throw e;
	}
	return Response.json(ids);
}
function handleLoopbackRequest(request, mf) {
	const url = new URL(request.url);
	if (url.pathname === "/snapshot") return handleSnapshotRequest(request, url);
	if (url.pathname === "/durable-objects") return listDurableObjectIds(request, mf, url);
	return new Response(null, { status: 404 });
}

//#endregion
//#region src/shared/builtin-modules.ts
const workerdBuiltinModules = new Set([...[
	"node-internal:async_hooks",
	"node-internal:buffer",
	"node-internal:crypto",
	"node-internal:module",
	"node-internal:process",
	"node-internal:util",
	"node-internal:diagnostics_channel",
	"node-internal:zlib",
	"node-internal:url",
	"node-internal:timers",
	"node-internal:sqlite",
	"node:_stream_duplex",
	"node:_stream_passthrough",
	"node:_stream_readable",
	"node:_stream_transform",
	"node:_stream_writable",
	"node:_tls_common",
	"node:_tls_wrap",
	"node:assert",
	"node:assert/strict",
	"node:async_hooks",
	"node:buffer",
	"node:constants",
	"node:crypto",
	"node:diagnostics_channel",
	"node:dns",
	"node:dns/promises",
	"node:events",
	"node:fs/promises",
	"node:module",
	"node:net",
	"node:path",
	"node:path/posix",
	"node:path/win32",
	"node:querystring",
	"node:stream",
	"node:stream/consumers",
	"node:stream/promises",
	"node:stream/web",
	"node:string_decoder",
	"node:test",
	"node:timers",
	"node:timers/promises",
	"node:tls",
	"node:url",
	"node:util",
	"node:util/types",
	"node:zlib",
	"node-internal:constants",
	"node-internal:crypto_cipher",
	"node-internal:crypto_dh",
	"node-internal:crypto_hash",
	"node-internal:crypto_hkdf",
	"node-internal:crypto_keys",
	"node-internal:crypto_pbkdf2",
	"node-internal:crypto_random",
	"node-internal:crypto_scrypt",
	"node-internal:crypto_sign",
	"node-internal:crypto_spkac",
	"node-internal:crypto_util",
	"node-internal:crypto_x509",
	"node-internal:debuglog",
	"node-internal:events",
	"node-internal:internal_assert",
	"node-internal:internal_assertionerror",
	"node-internal:internal_buffer",
	"node-internal:internal_comparisons",
	"node-internal:internal_diffs",
	"node-internal:internal_dns",
	"node-internal:internal_dns_client",
	"node-internal:internal_dns_constants",
	"node-internal:internal_dns_promises",
	"node-internal:internal_errors",
	"node-internal:internal_fs",
	"node-internal:internal_fs_callback",
	"node-internal:internal_fs_constants",
	"node-internal:internal_fs_promises",
	"node-internal:internal_fs_streams",
	"node-internal:internal_fs_sync",
	"node-internal:internal_fs_utils",
	"node-internal:internal_http",
	"node-internal:internal_http2_constants",
	"node-internal:internal_http_agent",
	"node-internal:internal_http_client",
	"node-internal:internal_http_constants",
	"node-internal:internal_http_incoming",
	"node-internal:internal_http_outgoing",
	"node-internal:internal_http_server",
	"node-internal:internal_http_util",
	"node-internal:internal_https_agent",
	"node-internal:internal_https_server",
	"node-internal:internal_inspect",
	"node-internal:internal_module",
	"node-internal:internal_net",
	"node-internal:internal_path",
	"node-internal:internal_process",
	"node-internal:internal_querystring",
	"node-internal:internal_readline",
	"node-internal:internal_readline_promises",
	"node-internal:internal_stringdecoder",
	"node-internal:internal_timers",
	"node-internal:internal_timers_global_override",
	"node-internal:internal_timers_promises",
	"node-internal:internal_tls",
	"node-internal:internal_tls_common",
	"node-internal:internal_tls_constants",
	"node-internal:internal_tls_jsstream",
	"node-internal:internal_tls_wrap",
	"node-internal:internal_types",
	"node-internal:internal_url",
	"node-internal:internal_utils",
	"node-internal:internal_zlib",
	"node-internal:internal_zlib_base",
	"node-internal:internal_zlib_constants",
	"node-internal:legacy_process",
	"node-internal:legacy_url",
	"node-internal:mock",
	"node-internal:public_process",
	"node-internal:streams_add_abort_signal",
	"node-internal:streams_compose",
	"node-internal:streams_destroy",
	"node-internal:streams_duplex",
	"node-internal:streams_end_of_stream",
	"node-internal:streams_legacy",
	"node-internal:streams_pipeline",
	"node-internal:streams_promises",
	"node-internal:streams_readable",
	"node-internal:streams_state",
	"node-internal:streams_transform",
	"node-internal:streams_util",
	"node-internal:streams_writable",
	"node-internal:validators",
	"internal:unsafe-eval",
	"cloudflare-internal:sockets",
	"cloudflare-internal:base64",
	"cloudflare-internal:messagechannel",
	"cloudflare:ai",
	"cloudflare:br",
	"cloudflare:email",
	"cloudflare:node",
	"cloudflare:pipelines",
	"cloudflare:sockets",
	"cloudflare:vectorize",
	"cloudflare:workers",
	"cloudflare:workflows",
	"cloudflare-internal:ai-api",
	"cloudflare-internal:aig-api",
	"cloudflare-internal:autorag-api",
	"cloudflare-internal:br-api",
	"cloudflare-internal:d1-api",
	"cloudflare-internal:http",
	"cloudflare-internal:images-api",
	"cloudflare-internal:pipeline-transform",
	"cloudflare-internal:streaming-base64",
	"cloudflare-internal:streaming-forms",
	"cloudflare-internal:test-tracing-wrapper",
	"cloudflare-internal:to-markdown-api",
	"cloudflare-internal:tracing-helpers",
	"cloudflare-internal:vectorize-api",
	"cloudflare-internal:workflows-api",
	"cloudflare-internal:workers",
	"cloudflare-internal:tracing",
	"cloudflare-internal:env",
	"cloudflare-internal:filesystem",
	"workerd:unsafe"
], "__STATIC_CONTENT_MANIFEST"]);

//#endregion
//#region src/pool/module-fallback.ts
let debuglog = util.debuglog("vitest-pool-workers:module-fallback", (log$2) => debuglog = log$2);
const isWindows = process.platform === "win32";
function ensurePosixLikePath(filePath) {
	return isWindows ? filePath.replaceAll("\\", "/") : filePath;
}
const __filename$1 = fileURLToPath(import.meta.url);
const __dirname$1 = path.dirname(__filename$1);
const require$1 = createRequire(__filename$1);
const distPath = ensurePosixLikePath(path.resolve(__dirname$1, ".."));
const libPath = posixPath.join(distPath, "worker", "lib");
const emptyLibPath = posixPath.join(libPath, "cloudflare/empty-internal.cjs");
const disableCjsEsmShimSuffix = "?mf_vitest_no_cjs_esm_shim";
function trimSuffix(suffix, value) {
	assert(value.endsWith(suffix));
	return value.substring(0, value.length - suffix.length);
}
/**
* When pre-bundling is enabled, Vite will add a hash to the end of the file path
* e.g. `/node_modules/.vite/deps/my-dep.js?v=f3sf2ebd`
*
* @see https://vite.dev/guide/features.html#npm-dependency-resolving-and-pre-bundling
* @see https://github.com/cloudflare/workers-sdk/pull/5673
*/
const versionHashRegExp = /\?v=[0-9a-f]+$/;
function trimViteVersionHash(filePath) {
	return filePath.replace(versionHashRegExp, "");
}
const forceModuleTypeRegexp = /* @__PURE__ */ new RegExp(`\\?mf_vitest_force=(${ModuleRuleTypeSchema.options.join("|")})$`);
function isFile(filePath) {
	return fs.statSync(filePath, { throwIfNoEntry: false })?.isFile() ?? false;
}
function isDirectory(filePath) {
	return fs.statSync(filePath, { throwIfNoEntry: false })?.isDirectory() ?? false;
}
function getParentPaths(filePath) {
	const parentPaths = [];
	while (true) {
		const parentPath = posixPath.dirname(filePath);
		if (parentPath === filePath) return parentPaths;
		parentPaths.push(parentPath);
		filePath = parentPath;
	}
}
const dirPathTypeModuleCache = /* @__PURE__ */ new Map();
function isWithinTypeModuleContext(filePath) {
	const parentPaths = getParentPaths(filePath);
	for (const parentPath of parentPaths) {
		const cache$1 = dirPathTypeModuleCache.get(parentPath);
		if (cache$1 !== void 0) return cache$1;
	}
	for (const parentPath of parentPaths) try {
		const pkgPath = posixPath.join(parentPath, "package.json");
		const pkgJson = fs.readFileSync(pkgPath, "utf8");
		const pkg = JSON.parse(pkgJson);
		const maybeModulePath = pkg.module ? posixPath.join(parentPath, pkg.module) : "";
		const cache$1 = pkg.type === "module" || maybeModulePath === filePath;
		dirPathTypeModuleCache.set(parentPath, cache$1);
		return cache$1;
	} catch (e) {
		if (!isFileNotFoundError(e)) throw e;
	}
	return false;
}
await cjsModuleLexer.init();
/**
* Gets "named" exports from a CommonJS module. Normally, CommonJS modules can
* only be default-imported, but Node performs additional static analysis to
* allow named-imports too (https://nodejs.org/api/esm.html#interoperability-with-commonjs).
* This function returns the named-exports we should add to our ESM-CJS shim,
* using the same package as Node.
*/
async function getCjsNamedExports(vite, filePath, contents, seen = /* @__PURE__ */ new Set()) {
	const { exports: exports$1, reexports } = cjsModuleLexer.parse(contents);
	const result = new Set(exports$1);
	for (const reexport of reexports) {
		const resolved = await viteResolve(vite, reexport, filePath, true);
		if (seen.has(resolved)) continue;
		try {
			const resolvedContents = fs.readFileSync(resolved, "utf8");
			seen.add(resolved);
			const resolvedNames = await getCjsNamedExports(vite, resolved, resolvedContents, seen);
			for (const name of resolvedNames) result.add(name);
		} catch (e) {
			if (!isFileNotFoundError(e)) throw e;
		}
	}
	result.delete("default");
	result.delete("__esModule");
	return result;
}
function withSourceUrl(contents, url) {
	if (contents.lastIndexOf("//# sourceURL=") !== -1) return contents;
	return contents + `\n//# sourceURL=${url.toString()}\n`;
}
function withImportMetaUrl(contents, url) {
	return contents.replaceAll("import.meta.url", JSON.stringify(url.toString()));
}
const moduleExtensions = [
	".js",
	".mjs",
	".cjs",
	".json"
];
function maybeGetTargetFilePath(target) {
	if (isFile(target)) return target;
	for (const extension of moduleExtensions) {
		const targetWithExtension = target + extension;
		if (fs.existsSync(targetWithExtension)) return targetWithExtension;
	}
	if (target.endsWith(disableCjsEsmShimSuffix)) return target;
	if (isDirectory(target)) return maybeGetTargetFilePath(target + "/index");
}
/**
* `target` is the path to the "file" `workerd` is trying to load,
* `referrer` is the path to the file that imported/required the `target`,
* `referrerDir` is the dirname of `referrer`
*
* For example, if the `referrer` is "/a/b/c/index.mjs":
*
* | Import Statement            | `target`           | Return             |
* |-----------------------------|--------------------|--------------------|
* | import "./dep.mjs"          | /a/b/c/dep.mjs     | dep.mjs            |
* | import "../dep.mjs"         | /a/b/dep.mjs       | ../dep.mjs         |
* | import "pkg"                | /a/b/c/pkg         | pkg                |
* | import "@org/pkg"           | /a/b/c/@org/pkg    | @org/pkg           |
* | import "node:assert"        | node:assert        | node:assert        |
* | import "cloudflare:sockets" | cloudflare:sockets | cloudflare:sockets |
* | import "workerd:rtti"       | workerd:rtti       | workerd:rtti       |
* | import "random:pkg"         | /a/b/c/random:pkg  | random:pkg         |
*
* Note that we return `dep.mjs` for `import "./dep.mjs"`. This would fail
* ES module resolution, so must be handled by `maybeGetTargetFilePath()`.
*/
function getApproximateSpecifier(target, referrerDir) {
	if (/^(node|cloudflare|workerd):/.test(target)) return target;
	return posixPath.relative(referrerDir, target);
}
async function viteResolve(vite, specifier, referrer, isRequire) {
	const resolved = await vite.pluginContainer.resolveId(specifier, referrer, {
		ssr: true,
		custom: { "node-resolve": { isRequire } }
	});
	if (resolved === null) {
		if (isRequire && specifier[0] === ".") return require$1.resolve(specifier, { paths: [referrer] });
		throw new Error("Not found");
	}
	if (resolved.id === "__vite-browser-external") return emptyLibPath;
	if (resolved.external) {
		let { id } = resolved;
		if (workerdBuiltinModules.has(id)) return `/${id}`;
		if (id.startsWith("node:")) throw new Error("Not found");
		id = `node:${id}`;
		if (workerdBuiltinModules.has(id)) return `/${id}`;
		return id;
	}
	return trimViteVersionHash(resolved.id);
}
async function resolve(vite, method, target, specifier, referrer) {
	const referrerDir = posixPath.dirname(referrer);
	let filePath = maybeGetTargetFilePath(target);
	if (filePath !== void 0) return filePath;
	if (referrerDir !== "/" && workerdBuiltinModules.has(specifier)) return `/${specifier}`;
	filePath = maybeGetTargetFilePath(posixPath.join(libPath, specifier.replaceAll(":", "/")));
	if (filePath !== void 0) return filePath;
	return viteResolve(vite, specifier, referrer, method === "require");
}
function buildRedirectResponse(filePath) {
	if (isWindows && filePath[0] !== "/") filePath = `/${filePath}`;
	return new Response(null, {
		status: 301,
		headers: { Location: filePath }
	});
}
function maybeGetForceTypeModuleContents(filePath) {
	const match = forceModuleTypeRegexp.exec(filePath);
	if (match === null) return;
	filePath = trimSuffix(match[0], filePath);
	const type = match[1];
	const contents = fs.readFileSync(filePath);
	switch (type) {
		case "ESModule": return { esModule: contents.toString() };
		case "CommonJS": return { commonJsModule: contents.toString() };
		case "Text": return { text: contents.toString() };
		case "Data": return { data: contents };
		case "CompiledWasm": return { wasm: contents };
		case "PythonModule": return { pythonModule: contents.toString() };
		case "PythonRequirement": return { pythonRequirement: contents.toString() };
		default: {
			const exhaustive = type;
			assert.fail(`Unreachable: ${exhaustive} modules are unsupported`);
		}
	}
}
function buildModuleResponse(target, contents) {
	let name = target;
	if (!isWindows) name = posixPath.relative("/", target);
	assert(name[0] !== "/");
	const result = { name };
	for (const key in contents) {
		const value = contents[key];
		result[key] = value instanceof Uint8Array ? Array.from(value) : value;
	}
	return Response.json(result);
}
async function load(vite, logBase, method, target, specifier, filePath) {
	if (target !== filePath) {
		if (method === "require" && !specifier.startsWith("node:")) filePath += disableCjsEsmShimSuffix;
		debuglog(logBase, "redirect:", filePath);
		return buildRedirectResponse(filePath);
	}
	if (filePath.endsWith(".wasm")) filePath += `?mf_vitest_force=CompiledWasm`;
	const maybeContents = maybeGetForceTypeModuleContents(filePath);
	if (maybeContents !== void 0) {
		debuglog(logBase, "forced:", filePath);
		return buildModuleResponse(target, maybeContents);
	}
	const disableCjsEsmShim = filePath.endsWith(disableCjsEsmShimSuffix);
	if (disableCjsEsmShim) filePath = trimSuffix(disableCjsEsmShimSuffix, filePath);
	const isEsm = filePath.endsWith(".mjs") || filePath.endsWith(".js") && isWithinTypeModuleContext(filePath);
	if (filePath.endsWith(".json")) {
		const json = fs.readFileSync(filePath, "utf8");
		debuglog(logBase, "json:", filePath);
		return buildModuleResponse(target, { json });
	}
	let contents = fs.readFileSync(filePath, "utf8");
	const targetUrl = pathToFileURL(target);
	contents = withSourceUrl(contents, targetUrl);
	if (isEsm) {
		contents = withImportMetaUrl(contents, targetUrl);
		debuglog(logBase, "esm:", filePath);
		return buildModuleResponse(target, { esModule: contents });
	}
	if ((method === "import" || specifier.startsWith("node:")) && !disableCjsEsmShim) {
		const disableShimSpecifier = `./${posixPath.basename(filePath)}${disableCjsEsmShimSuffix}`;
		let esModule = `import mod from ${JSON.stringify(disableShimSpecifier)}; export default mod;`;
		for (const name of await getCjsNamedExports(vite, filePath, contents)) esModule += ` export const ${name} = mod.${name};`;
		debuglog(logBase, "cjs-esm-shim:", filePath);
		return buildModuleResponse(target, { esModule });
	}
	debuglog(logBase, "cjs:", filePath);
	return buildModuleResponse(target, { commonJsModule: contents });
}
async function handleModuleFallbackRequest(vite, request) {
	const method = request.headers.get("X-Resolve-Method");
	assert(method === "import" || method === "require");
	const url = new URL(request.url);
	let target = url.searchParams.get("specifier");
	let referrer = url.searchParams.get("referrer");
	assert(target !== null, "Expected specifier search param");
	assert(referrer !== null, "Expected referrer search param");
	const referrerDir = posixPath.dirname(referrer);
	let specifier = getApproximateSpecifier(target, referrerDir);
	if (specifier.startsWith("file:")) specifier = fileURLToPath(specifier);
	if (isWindows) {
		if (target[0] === "/") target = target.substring(1);
		if (referrer[0] === "/") referrer = referrer.substring(1);
	}
	const logBase = `${method}(${JSON.stringify(target)}) relative to ${referrer}:`;
	try {
		const filePath = await resolve(vite, method, target, specifier, referrer);
		return await load(vite, logBase, method, target, specifier, filePath);
	} catch (e) {
		debuglog(logBase, "error:", e);
		console.error(`[vitest-pool-workers] Failed to ${method} ${JSON.stringify(target)} from ${JSON.stringify(referrer)}.`, "To resolve this, try bundling the relevant dependency with Vite.", "For more details, refer to https://developers.cloudflare.com/workers/testing/vitest-integration/known-issues/#module-resolution");
	}
	return new Response(null, { status: 404 });
}

//#endregion
//#region src/pool/config.ts
const PLUGIN_VALUES = Object.values(PLUGINS);
const WorkersPoolOptionsSchema = z.object({
	main: z.ostring(),
	remoteBindings: z.boolean().default(true),
	additionalExports: z.record(z.string(), z.union([
		z.literal("WorkerEntrypoint"),
		z.literal("DurableObject"),
		z.literal("WorkflowEntrypoint")
	])).default({}),
	miniflare: z.object({ workers: z.array(z.object({}).passthrough()).optional() }).passthrough().optional(),
	wrangler: z.object({
		configPath: z.ostring(),
		environment: z.ostring()
	}).optional()
});
function isZodErrorLike(value) {
	return typeof value === "object" && value !== null && "issues" in value && Array.isArray(value.issues);
}
function coalesceZodErrors(ref, thrown) {
	if (!isZodErrorLike(thrown)) throw thrown;
	if (ref.value === void 0) ref.value = thrown;
	else ref.value.issues.push(...thrown.issues);
}
function parseWorkerOptions(rootPath, value, withoutScript, opts) {
	if (withoutScript) {
		value["script"] = "";
		delete value["scriptPath"];
		delete value["modules"];
		delete value["modulesRoot"];
	}
	const result = {};
	const errorRef = {};
	for (const plugin of PLUGIN_VALUES) try {
		const parsed = parseWithRootPath(rootPath, plugin.options, value, opts);
		Object.assign(result, parsed);
	} catch (e) {
		coalesceZodErrors(errorRef, e);
	}
	if (errorRef.value !== void 0) throw errorRef.value;
	if (withoutScript) delete value["script"];
	return result;
}
const log$1 = new Log(LogLevel.WARN, { prefix: "vpw" });
function filterTails(tails, userWorkers) {
	return tails?.filter((tailService) => {
		let name;
		if (typeof tailService === "string") name = tailService;
		else if (typeof tailService === "object" && "name" in tailService && typeof tailService.name === "string") name = tailService.name;
		else return true;
		const found = userWorkers?.some((w) => w.name === name);
		if (!found) log$1.warn(`Tail consumer "${name}" was not found in your config. Make sure you add it if you'd like to simulate receiving tail events locally.`);
		return found;
	});
}
/** Map that maps worker configPaths to their existing remote proxy session data (if any) */
const remoteProxySessionsDataMap = /* @__PURE__ */ new Map();
async function parseCustomPoolOptions(rootPath, value) {
	const options = WorkersPoolOptionsSchema.parse(value);
	options.miniflare ??= {};
	const errorRef = {};
	const workers = options.miniflare?.workers;
	const rootPathOption = getRootPath(options.miniflare);
	rootPath = path.resolve(rootPath, rootPathOption);
	try {
		options.miniflare = parseWorkerOptions(rootPath, options.miniflare, true, { path: ["miniflare"] });
	} catch (e) {
		coalesceZodErrors(errorRef, e);
	}
	options.miniflare.workers = [];
	if (workers !== void 0) options.miniflare.workers = workers.map((worker, i) => {
		try {
			const workerRootPathOption = getRootPath(worker);
			return parseWorkerOptions(path.resolve(rootPath, workerRootPathOption), worker, false, { path: [
				"miniflare",
				"workers",
				i
			] });
		} catch (e) {
			coalesceZodErrors(errorRef, e);
			return { script: "" };
		}
	});
	if (errorRef.value !== void 0) throw errorRef.value;
	if (options.wrangler?.configPath !== void 0) {
		const configPath = path.resolve(rootPath, options.wrangler.configPath);
		options.wrangler.configPath = configPath;
		const wrangler = await import("wrangler");
		const preExistingRemoteProxySessionData = options.wrangler?.configPath ? remoteProxySessionsDataMap.get(options.wrangler.configPath) : void 0;
		const remoteProxySessionData = options.remoteBindings ? await wrangler.maybeStartOrUpdateRemoteProxySession({
			path: options.wrangler.configPath,
			environment: options.wrangler.environment
		}, preExistingRemoteProxySessionData ?? null) : null;
		if (options.wrangler?.configPath && remoteProxySessionData) remoteProxySessionsDataMap.set(options.wrangler.configPath, remoteProxySessionData);
		const { workerOptions, externalWorkers, define, main } = wrangler.unstable_getMiniflareWorkerOptions(configPath, options.wrangler.environment, {
			overrides: {
				assets: options.miniflare.assets,
				enableContainers: false
			},
			remoteProxyConnectionString: remoteProxySessionData?.session?.remoteProxyConnectionString
		});
		options.main ??= main;
		options.miniflare.workers = [...options.miniflare.workers, ...externalWorkers];
		options.miniflare = mergeWorkerOptions(workerOptions, options.miniflare);
		options.miniflare = {
			...options.miniflare,
			tails: filterTails(workerOptions.tails, options.miniflare.workers)
		};
		options.defines = define;
	}
	if (options.miniflare?.assets) {
		options.miniflare.hasAssetsAndIsVitest = true;
		options.miniflare.assets.routerConfig ??= {};
		options.miniflare.assets.routerConfig.has_user_worker = Boolean(options.main);
	}
	return options;
}
async function parseProjectOptions(project, poolOptions) {
	const environment = project.config.environment;
	if (environment !== void 0 && environment !== "node") {
		const quotedEnvironment = JSON.stringify(environment);
		let migrationGuide = ".";
		if (environment === "miniflare") migrationGuide = ", and refer to the migration guide if upgrading from `vitest-environment-miniflare`:\nhttps://developers.cloudflare.com/workers/testing/vitest-integration/get-started/migrate-from-miniflare-2/";
		const message = [
			`Unexpected custom \`environment\` ${quotedEnvironment} in project ${getRelativeProjectPath(project)}.`,
			"The Workers pool always runs your tests inside of an environment providing Workers runtime APIs.",
			`Please remove the \`environment\` configuration${migrationGuide}`
		].join("\n");
		throw new TypeError(message);
	}
	const projectPath = getProjectPath(project);
	try {
		return await parseCustomPoolOptions(projectPath, poolOptions);
	} catch (e) {
		if (!isZodErrorLike(e)) throw e;
		let formatted;
		try {
			formatted = formatZodError(e, poolOptions);
		} catch {
			throw e;
		}
		const relativePath = getRelativeProjectConfigPath(project);
		throw new TypeError(`Unexpected options in project ${relativePath}:\n${formatted}`);
	}
}

//#endregion
//#region src/pool/pages.ts
const registeredControllers = /* @__PURE__ */ new Set();
let activeWorkers = 0;
function poolWorkerStarted() {
	activeWorkers++;
}
function poolWorkerStopped() {
	activeWorkers--;
	if (activeWorkers <= 0) {
		activeWorkers = 0;
		for (const ac of registeredControllers) ac.abort();
		registeredControllers.clear();
	}
}
async function buildPagesASSETSBinding(assetsPath) {
	if (typeof assetsPath !== "string") throw new TypeError("Failed to execute 'buildPagesASSETSBinding': parameter 1 is not of type 'string'.");
	const { unstable_generateASSETSBinding } = await import("wrangler");
	const ac = new AbortController();
	registeredControllers.add(ac);
	return unstable_generateASSETSBinding({
		log: {
			...console,
			debugWithSanitization: console.debug,
			loggerLevel: "info",
			columns: process.stdout.columns
		},
		directory: assetsPath,
		signal: ac.signal
	});
}

//#endregion
//#region src/pool/cloudflare-pool-worker.ts
var CloudflarePoolWorker = class {
	name = "cloudflare-pool";
	mf;
	socket;
	parsedPoolOptions;
	main;
	messageListener;
	errorListener;
	closeListener;
	constructor(options, poolOptions) {
		this.options = options;
		this.poolOptions = poolOptions;
		assertCompatibleVitestVersion(options.project.vitest);
	}
	async start() {
		poolWorkerStarted();
		let resolvedPoolOptions;
		if (typeof this.poolOptions === "function") {
			const inject = (key) => {
				return this.options.project.getProvidedContext()[key];
			};
			resolvedPoolOptions = await this.poolOptions({ inject });
		} else resolvedPoolOptions = this.poolOptions;
		this.parsedPoolOptions = await parseProjectOptions(this.options.project, resolvedPoolOptions);
		this.main = maybeGetResolvedMainPath(this.options.project, this.parsedPoolOptions);
		const configPlugin = this.options.project.vite.config.plugins.find(({ name }) => name === "@cloudflare/vitest-pool-workers");
		if (configPlugin !== void 0) configPlugin.api.setMain(this.main);
		this.mf = await getProjectMiniflare(this.options.project.vitest, this.options.project, this.parsedPoolOptions, this.main);
		this.socket = await connectToMiniflareSocket(this.mf, getRunnerName(this.options.project));
	}
	async stop() {
		this.socket?.close();
		this.socket = void 0;
		await this.mf?.dispose();
		this.mf = void 0;
		if (this.parsedPoolOptions?.wrangler?.configPath) await remoteProxySessionsDataMap.get(this.parsedPoolOptions?.wrangler?.configPath)?.session?.dispose?.();
		poolWorkerStopped();
	}
	send(message) {
		assert(this.socket, "Message sent to Worker before initialisation");
		assert(this.parsedPoolOptions, "Message sent to Worker before initialisation");
		let toSend = message;
		if (message.type === "start") toSend = {
			...message,
			context: {
				...message.context,
				config: {
					...message.context.config,
					inspector: {
						...message.context.config.inspector,
						enabled: false
					}
				}
			}
		};
		else if (message.type === "run") toSend = {
			...message,
			context: {
				...message.context,
				providedContext: {
					...message.context.providedContext,
					cloudflarePoolOptions: JSON.stringify({
						main: this.main,
						durableObjectBindingDesignators: [...getDurableObjectDesignators(this.parsedPoolOptions).entries()],
						selfName: getRunnerName(this.options.project)
					})
				}
			}
		};
		this.socket.send(structuredSerializableStringify(toSend));
	}
	on(event, callback) {
		assert(this.socket, "Message received from Worker before initialisation");
		assert(this.parsedPoolOptions, "Message received from Worker before initialisation");
		const rules = this.parsedPoolOptions.miniflare?.modulesRules;
		const compiledRules = compileModuleRules(rules ?? []);
		if (event === "message") {
			const messageWrapper = (m) => {
				const d = structuredSerializableParse(m.data);
				if (d && typeof d === "object" && "m" in d && "a" in d && "i" in d && Array.isArray(d.a) && d.m === "fetch") {
					assert(this.socket, "Message received from Worker before initialisation");
					const specifier = d.a[0];
					if (specifier !== "cloudflare:test" && (/^(cloudflare|workerd):/.test(specifier) || workerdBuiltinModules.has(specifier))) return this.socket.send(structuredSerializableStringify({
						t: "s",
						i: d.i,
						r: { externalize: specifier }
					}));
					const maybeRule = compiledRules.find((rule) => testRegExps(rule.include, specifier));
					if (maybeRule !== void 0 && !specifier.includes("?")) {
						const externalize = path.join(this.options.project.config.root, specifier) + `?mf_vitest_force=${maybeRule.type}`;
						return this.socket.send(structuredSerializableStringify({
							t: "s",
							i: d.i,
							r: { externalize }
						}));
					}
				}
				callback(d);
			};
			this.messageListener = messageWrapper;
			this.socket.addEventListener("message", this.messageListener);
		} else if (event === "error") {
			this.errorListener = (e) => {
				callback("error" in e ? e.error : e);
			};
			this.socket.addEventListener("error", this.errorListener);
		} else if (event === "exit") {
			this.closeListener = callback;
			this.socket.addEventListener("close", this.closeListener);
		}
	}
	off(event, _callback) {
		if (event === "message" && this.messageListener) {
			this.socket?.removeEventListener("message", this.messageListener);
			this.messageListener = void 0;
		} else if (event === "error" && this.errorListener) {
			this.socket?.removeEventListener("error", this.errorListener);
			this.errorListener = void 0;
		} else if (event === "exit" && this.closeListener) {
			this.socket?.removeEventListener("close", this.closeListener);
			this.closeListener = void 0;
		}
	}
	deserialize(data) {
		return data;
	}
};

//#endregion
//#region src/pool/pool.ts
function cloudflarePool(poolOptions) {
	return {
		name: "cloudflare-pool",
		createPoolWorker: (options) => new CloudflarePoolWorker(options, poolOptions)
	};
}

//#endregion
//#region src/pool/plugin.ts
const cloudflareTestPath = path.resolve(import.meta.dirname, "../worker/lib/cloudflare/test.mjs");
function ensureArrayIncludes(array, items) {
	for (const item of items) if (!array.includes(item)) array.push(item);
}
function ensureArrayExcludes(array, items) {
	for (let i = 0; i < array.length; i++) if (items.includes(array[i])) {
		array.splice(i, 1);
		i--;
	}
}
const requiredConditions = [
	"workerd",
	"worker",
	"module",
	"browser"
];
const requiredMainFields = [
	"browser",
	"module",
	"jsnext:main",
	"jsnext"
];
function cloudflareTest(options) {
	const uuid = crypto.randomUUID();
	let main;
	return {
		name: "@cloudflare/vitest-pool-workers",
		api: { setMain(newMain) {
			main = newMain;
		} },
		configureVitest(context) {
			context.project.config.poolRunner = cloudflarePool(options);
			context.project.config.pool = "cloudflare-pool";
			context.project.config.snapshotEnvironment = "cloudflare:snapshot";
		},
		config(config) {
			config.resolve ??= {};
			config.resolve.conditions ??= [];
			config.resolve.mainFields ??= [];
			config.ssr ??= {};
			config.test ??= {};
			config.test.server ??= {};
			config.test.server.deps ??= {};
			config.test.server.deps.inline = true;
			ensureArrayExcludes(config.resolve.conditions, ["node"]);
			ensureArrayIncludes(config.resolve.conditions, requiredConditions);
			ensureArrayIncludes(config.resolve.mainFields, requiredMainFields);
			config.ssr.target = "webworker";
		},
		resolveId(id) {
			if (id === "cloudflare:test") return `\0cloudflare:test-${uuid}`;
		},
		async load(id) {
			if (id === `\0cloudflare:test-${uuid}`) {
				let contents = await fs$1.readFile(cloudflareTestPath, "utf8");
				if (main !== void 0) contents += `import ${JSON.stringify(main)};`;
				return contents;
			}
			if (id.endsWith("msw/lib/node/index.mjs")) return `export * from "../native/index.mjs"`;
		}
	};
}

//#endregion
//#region src/pool/d1.ts
/**
* Reads all migrations in `migrationsPath`, ordered by migration number.
* Each migration will have its contents split into an array of SQL queries.
*/
async function readD1Migrations(migrationsPath) {
	if (typeof migrationsPath !== "string") throw new TypeError("Failed to execute 'readD1Migrations': parameter 1 is not of type 'string'.");
	const { unstable_splitSqlQuery } = await import("wrangler");
	const names = fs.readdirSync(migrationsPath).filter((name) => name.endsWith(".sql"));
	names.sort((a, b) => {
		return parseInt(a.split("_")[0]) - parseInt(b.split("_")[0]);
	});
	return names.map((name) => {
		const migrationPath = path.join(migrationsPath, name);
		return {
			name,
			queries: unstable_splitSqlQuery(fs.readFileSync(migrationPath, "utf8"))
		};
	});
}

//#endregion
//#region src/pool/index.ts
function structuredSerializableStringify(value) {
	if (value && typeof value === "object" && "r" in value && value.r && typeof value.r === "object" && "map" in value.r && value.r.map) value = {
		...value,
		r: {
			...value.r,
			map: void 0
		}
	};
	return stringify(value, structuredSerializableReducers);
}
function structuredSerializableParse(value) {
	return parse(value, structuredSerializableRevivers);
}
const log = new Log(LogLevel.VERBOSE, { prefix: "vpw" });
const mfLog = new Log(LogLevel.WARN);
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const DIST_PATH = path.resolve(__dirname, "..");
const POOL_WORKER_PATH = path.join(DIST_PATH, "worker/index.mjs");
const symbolizerWarning = "warning: Not symbolizing stack traces because $LLVM_SYMBOLIZER is not set.";
const ignoreMessages = [
	"disconnected: operation canceled",
	"disconnected: worker_do_not_log; Request failed due to internal error",
	"disconnected: WebSocket was aborted"
];
function trimSymbolizerWarning(chunk) {
	return chunk.includes(symbolizerWarning) ? chunk.substring(chunk.indexOf("\n") + 1) : chunk;
}
function handleRuntimeStdio(stdout, stderr) {
	stdout.on("data", (chunk) => {
		process.stdout.write(chunk);
	});
	stderr.on("data", (chunk) => {
		const str = trimSymbolizerWarning(chunk.toString());
		if (ignoreMessages.some((message) => str.includes(message))) return;
		process.stderr.write(str);
	});
}
function getRunnerName(project, testFile) {
	const name = `${WORKER_NAME_PREFIX}runner-${project.name.replace(/[^a-z0-9-]/gi, "_")}`;
	if (testFile === void 0) return name;
	const testFileHash = crypto.createHash("sha1").update(testFile).digest("hex");
	testFile = testFile.replace(/[^a-z0-9-]/gi, "_");
	return `${name}-${testFileHash}-${testFile}`;
}
function isDurableObjectDesignatorToSelf(value) {
	if (typeof value === "string") return true;
	return typeof value === "object" && value !== null && "className" in value && typeof value.className === "string" && (!("scriptName" in value) || value.scriptName === void 0);
}
function isWorkflowDesignatorToSelf(value, currentScriptName) {
	return typeof value === "object" && value !== null && "className" in value && typeof value.className === "string" && (!("scriptName" in value) || value.scriptName === void 0 || value.scriptName === currentScriptName);
}
/**
* Returns a map of Durable Objects bindings' bound names to the designators of
* the objects they point to.
*/
function getDurableObjectDesignators(options) {
	const result = /* @__PURE__ */ new Map();
	const durableObjects = options.miniflare?.durableObjects ?? {};
	for (const [key, designator] of Object.entries(durableObjects)) if (typeof designator === "string") result.set(key, { className: designator });
	else if (typeof designator.unsafeUniqueKey !== "symbol") result.set(key, {
		className: designator.className,
		scriptName: designator.scriptName,
		unsafeUniqueKey: designator.unsafeUniqueKey
	});
	for (const unboundDurableObject of options.miniflare?.additionalUnboundDurableObjects ?? []) if (typeof unboundDurableObject.unsafeUniqueKey !== "symbol") result.set(unboundDurableObject.className, {
		className: unboundDurableObject.className,
		scriptName: unboundDurableObject.scriptName,
		unsafeUniqueKey: unboundDurableObject.unsafeUniqueKey
	});
	return result;
}
/**
* Gets a set of Durable Object class names for the SELF Worker.
*
* This is calculated from the Durable Object bindings that point to SELF as well as the
* unbound Durable Objects that only have migrations defined.
*/
function getDurableObjectClasses(worker) {
	const result = /* @__PURE__ */ new Set();
	for (const designator of Object.values(worker.durableObjects ?? {})) if (isDurableObjectDesignatorToSelf(designator)) result.add(typeof designator === "string" ? designator : designator.className);
	for (const designator of worker.additionalUnboundDurableObjects ?? []) result.add(designator.className);
	return result;
}
function getWranglerWorkerName(relativeWranglerConfigPath) {
	if (!relativeWranglerConfigPath) return;
	return experimental_readRawConfig({ config: relativeWranglerConfigPath }).rawConfig.name;
}
/**
* Gets a set of class names for Workflows defined in the SELF Worker.
*/
function getWorkflowClasses(worker, relativeWranglerConfigPath) {
	const result = /* @__PURE__ */ new Set();
	if (worker.workflows === void 0) return result;
	for (const key of Object.keys(worker.workflows)) {
		const designator = worker.workflows[key];
		let workerName;
		const wranglerWorkerName = getWranglerWorkerName(relativeWranglerConfigPath);
		if (wranglerWorkerName && designator.scriptName === wranglerWorkerName) workerName = wranglerWorkerName;
		else workerName = worker.name;
		if (isWorkflowDesignatorToSelf(designator, workerName)) {
			result.add(designator.className);
			worker.workflows[key] = { ...designator };
		}
	}
	return result;
}
const SELF_SERVICE_BINDING = "__VITEST_POOL_WORKERS_SELF_SERVICE";
const LOOPBACK_SERVICE_BINDING = "__VITEST_POOL_WORKERS_LOOPBACK_SERVICE";
const RUNNER_OBJECT_BINDING = "__VITEST_POOL_WORKERS_RUNNER_OBJECT";
async function buildProjectWorkerOptions(project, customOptions, main) {
	const relativeWranglerConfigPath = maybeApply((v) => path.relative("", v), customOptions.wrangler?.configPath);
	const runnerWorker = customOptions.miniflare ?? {};
	runnerWorker.name = getRunnerName(project);
	runnerWorker.compatibilityFlags ??= [];
	runnerWorker.compatibilityFlags.push("no_handle_cross_request_promise_resolution");
	if (runnerWorker.compatibilityDate === void 0) {
		runnerWorker.compatibilityDate ??= supportedCompatibilityDate;
		log.info(`No compatibility date was provided for project ${getRelativeProjectPath(project)}, defaulting to latest supported date ${runnerWorker.compatibilityDate}.`);
	}
	const flagAssertions = new CompatibilityFlagAssertions({
		compatibilityDate: runnerWorker.compatibilityDate,
		compatibilityFlags: runnerWorker.compatibilityFlags,
		optionsPath: `miniflare`,
		relativeProjectPath: getRelativeProjectPath(project),
		relativeWranglerConfigPath
	});
	const assertions = [() => flagAssertions.assertIsEnabled({
		enableFlag: "export_commonjs_default",
		disableFlag: "export_commonjs_namespace",
		defaultOnDate: "2022-10-31"
	})];
	for (const assertion of assertions) {
		const result = assertion();
		if (!result.isValid) throw new Error(result.errorMessage);
	}
	const { hasNoNodejsCompatV2Flag, mode } = getNodeCompat(runnerWorker.compatibilityDate, runnerWorker.compatibilityFlags);
	if (mode !== "v2") {
		if (hasNoNodejsCompatV2Flag) runnerWorker.compatibilityFlags.splice(runnerWorker.compatibilityFlags.indexOf("no_nodejs_compat_v2"), 1);
		runnerWorker.compatibilityFlags.push("nodejs_compat_v2");
	}
	if (!runnerWorker.compatibilityFlags.includes("unsafe_module")) runnerWorker.compatibilityFlags.push("unsafe_module");
	ensureFeature(runnerWorker.compatibilityFlags, "nodejs_tty_module");
	ensureFeature(runnerWorker.compatibilityFlags, "nodejs_fs_module");
	ensureFeature(runnerWorker.compatibilityFlags, "nodejs_http_modules");
	ensureFeature(runnerWorker.compatibilityFlags, "nodejs_perf_hooks_module");
	ensureFeature(runnerWorker.compatibilityFlags, "nodejs_v8_module");
	ensureFeature(runnerWorker.compatibilityFlags, "nodejs_process_v2");
	runnerWorker.unsafeEvalBinding = "__VITEST_POOL_WORKERS_UNSAFE_EVAL";
	runnerWorker.unsafeUseModuleFallbackService = true;
	runnerWorker.serviceBindings ??= {};
	runnerWorker.serviceBindings[SELF_SERVICE_BINDING] = kCurrentWorker;
	runnerWorker.serviceBindings[LOOPBACK_SERVICE_BINDING] = handleLoopbackRequest;
	runnerWorker.durableObjects ??= {};
	const durableObjectClassNames = getDurableObjectClasses(runnerWorker);
	const workflowClassNames = getWorkflowClasses(runnerWorker, relativeWranglerConfigPath);
	const selfWorkerExports = [];
	if (flagAssertions.isEnabled("enable_ctx_exports", "disable_ctx_exports", "2025-11-17")) try {
		const guessedExports = await guessWorkerExports(main, customOptions.additionalExports);
		for (const [exportName, exportType] of guessedExports) switch (exportType) {
			case "DurableObject":
				durableObjectClassNames.add(exportName);
				break;
			case "WorkflowEntrypoint":
				workflowClassNames.add(exportName);
				break;
			case "WorkerEntrypoint":
			case null: selfWorkerExports.push(exportName);
		}
	} catch (e) {
		const message = `Failed to statically analyze the exports of the main Worker entry-point "${customOptions.main}"\nMore details: ${e}`;
		for (const line of message.split("\n")) log.warn(line);
	}
	const workerEntrypointExports = selfWorkerExports.filter((name) => name !== "default" && name !== "__esModule" && !durableObjectClassNames.has(name) && !workflowClassNames.has(name));
	const wrappers = ["import { createWorkerEntrypointWrapper, createDurableObjectWrapper, createWorkflowEntrypointWrapper } from \"cloudflare:test-internal\";"];
	for (const entrypointName of workerEntrypointExports.sort()) {
		const wrapper = `export const ${entrypointName} = createWorkerEntrypointWrapper(${JSON.stringify(entrypointName)});`;
		wrappers.push(wrapper);
	}
	for (const className of Array.from(durableObjectClassNames).sort()) {
		const wrapper = `export const ${className} = createDurableObjectWrapper(${JSON.stringify(className)});`;
		wrappers.push(wrapper);
	}
	for (const className of Array.from(workflowClassNames).sort()) {
		const wrapper = `export const ${className} = createWorkflowEntrypointWrapper(${JSON.stringify(className)});`;
		wrappers.push(wrapper);
	}
	runnerWorker.durableObjects[RUNNER_OBJECT_BINDING] = {
		className: "__VITEST_POOL_WORKERS_RUNNER_DURABLE_OBJECT__",
		unsafePreventEviction: true,
		unsafeUniqueKey: kUnsafeEphemeralUniqueKey
	};
	const defines = `export default {
		${Object.entries(customOptions.defines ?? {}).map(([key, value]) => `${JSON.stringify(key)}: ${value}`).join(",\n")}
	};
	`;
	if ("script" in runnerWorker) delete runnerWorker.script;
	if ("scriptPath" in runnerWorker) delete runnerWorker.scriptPath;
	const modulesRoot = process.platform === "win32" ? "Z:\\" : "/";
	runnerWorker.modulesRoot = modulesRoot;
	runnerWorker.modules = [
		{
			type: "ESModule",
			path: path.join(modulesRoot, "index.mjs"),
			contents: fs.readFileSync(POOL_WORKER_PATH)
		},
		{
			type: "ESModule",
			path: path.join(modulesRoot, "__VITEST_POOL_WORKERS_USER_OBJECT"),
			contents: wrappers.join("\n")
		},
		{
			type: "ESModule",
			path: path.join(modulesRoot, "__VITEST_POOL_WORKERS_DEFINES"),
			contents: defines
		},
		{
			type: "ESModule",
			path: path.join(modulesRoot, "node:console"),
			contents: fs.readFileSync(path.join(DIST_PATH, `worker/node/console.mjs`))
		},
		{
			type: "ESModule",
			path: path.join(modulesRoot, "node:vm"),
			contents: fs.readFileSync(path.join(DIST_PATH, `worker/node/vm.mjs`))
		}
	];
	const workers = [runnerWorker];
	if (runnerWorker.workers !== void 0) {
		for (let i = 0; i < runnerWorker.workers.length; i++) {
			const worker = runnerWorker.workers[i];
			if (typeof worker !== "object" || worker === null || !("name" in worker) || typeof worker.name !== "string" || worker.name === "") throw new Error(`In project ${getRelativeProjectPath(project)}, \`miniflare.workers[${i}].name\` must be non-empty`);
			if (worker.name.startsWith(WORKER_NAME_PREFIX)) throw new Error(`In project ${getRelativeProjectPath(project)}, \`miniflare.workers[${i}].name\` must not start with "${WORKER_NAME_PREFIX}", got ${worker.name}`);
			workers.push(worker);
		}
		delete runnerWorker.workers;
	}
	return workers;
}
const SHARED_MINIFLARE_OPTIONS = {
	log: mfLog,
	verbose: true,
	handleRuntimeStdio,
	unsafeStickyBlobs: true
};
const DEFAULT_INSPECTOR_PORT = 9229;
function getFirstAvailablePort(start) {
	return getPorts({ port: portNumbers(start, 65535) });
}
const moduleFallbackServices = /* @__PURE__ */ new WeakMap();
function getModuleFallbackService(ctx) {
	let service = moduleFallbackServices.get(ctx);
	if (service !== void 0) return service;
	service = handleModuleFallbackRequest.bind(void 0, ctx.vite);
	moduleFallbackServices.set(ctx, service);
	return service;
}
/**
* Builds options for the Miniflare instance running tests for the given Vitest
* project.
*/
async function buildProjectMiniflareOptions(ctx, project, customOptions, main) {
	const moduleFallbackService = getModuleFallbackService(ctx);
	const [runnerWorker, ...auxiliaryWorkers] = await buildProjectWorkerOptions(project, customOptions, main);
	assert(runnerWorker.name !== void 0);
	assert(runnerWorker.name.startsWith(WORKER_NAME_PREFIX));
	let inspectorPort;
	if (ctx.config.inspector.enabled) {
		const userSpecifiedPort = ctx.config.inspector.port;
		if (userSpecifiedPort !== void 0) {
			if (await getFirstAvailablePort(userSpecifiedPort) !== userSpecifiedPort) throw new Error(`Inspector port ${userSpecifiedPort} is not available. Either free up the port or remove the inspector port configuration to use an automatically assigned port.`);
			inspectorPort = userSpecifiedPort;
		} else {
			inspectorPort = await getFirstAvailablePort(DEFAULT_INSPECTOR_PORT);
			if (inspectorPort !== DEFAULT_INSPECTOR_PORT) log.warn(`Default inspector port ${DEFAULT_INSPECTOR_PORT} not available, using ${inspectorPort} instead.`);
		}
	}
	return {
		...SHARED_MINIFLARE_OPTIONS,
		inspectorPort,
		unsafeModuleFallbackService: moduleFallbackService,
		workers: [runnerWorker, ...auxiliaryWorkers]
	};
}
async function getProjectMiniflare(ctx, project, poolOptions, main) {
	const mfOptions = await buildProjectMiniflareOptions(ctx, project, poolOptions, main);
	log.info(`Starting runtime for ${getRelativeProjectPath(project)}${mfOptions.inspectorPort !== void 0 ? ` with inspector on port ${mfOptions.inspectorPort}` : ""}...`);
	const mf = new Miniflare(mfOptions);
	await mf.ready;
	return mf;
}
function maybeGetResolvedMainPath(project, options) {
	const projectPath = getProjectPath(project);
	const main = options.main;
	if (main === void 0) return;
	if (typeof projectPath === "string") return path.resolve(projectPath, main);
	else return path.resolve(main);
}
async function connectToMiniflareSocket(mf, workerName) {
	const res = await (await mf.getDurableObjectNamespace(RUNNER_OBJECT_BINDING, workerName)).get("singleton").fetch("http://placeholder", { headers: {
		Upgrade: "websocket",
		"MF-Vitest-Worker-Data": structuredSerializableStringify({ cwd: process.cwd() })
	} });
	const webSocket = res.webSocket;
	if (webSocket === null) {
		const body = await res.text().catch(() => "");
		throw new Error(`Failed to establish WebSocket to runner (status ${res.status}): ${body}`);
	}
	webSocket.accept();
	return webSocket;
}
function getPackageJson(dirPath) {
	while (true) {
		const pkgJsonPath = path.join(dirPath, "package.json");
		try {
			const contents = fs.readFileSync(pkgJsonPath, "utf8");
			return JSON.parse(contents);
		} catch (e) {
			if (!isFileNotFoundError(e)) throw e;
		}
		const nextDirPath = path.dirname(dirPath);
		if (nextDirPath === dirPath) return;
		dirPath = nextDirPath;
	}
}
function assertCompatibleVitestVersion(ctx) {
	const poolPkgJson = getPackageJson(__dirname);
	const vitestPkgJson = getPackageJson(ctx.distPath);
	assert(poolPkgJson !== void 0, "Expected to find `package.json` for `@cloudflare/vitest-pool-workers`");
	assert(vitestPkgJson !== void 0, "Expected to find `package.json` for `vitest`");
	const expectedVitestVersion = poolPkgJson.peerDependencies?.vitest;
	const actualVitestVersion = vitestPkgJson.version;
	assert(expectedVitestVersion !== void 0, "Expected to find `@cloudflare/vitest-pool-workers`'s `vitest` version constraint");
	assert(actualVitestVersion !== void 0, "Expected to find `vitest`'s version");
	if ((0, import_satisfies.default)(actualVitestVersion, "3.x")) {
		const message = `You're running \`vitest@${actualVitestVersion}\`, but this version of \`@cloudflare/vitest-pool-workers\` only supports \`vitest ${expectedVitestVersion}\`.`;
		throw new Error(message);
	}
	if (!(0, import_satisfies.default)(actualVitestVersion, expectedVitestVersion)) {
		const message = [
			`You're running \`vitest@${actualVitestVersion}\`, but this version of \`@cloudflare/vitest-pool-workers\` only officially supports \`vitest ${expectedVitestVersion}\`.`,
			"`@cloudflare/vitest-pool-workers` currently depends on internal Vitest APIs that are not protected by semantic-versioning guarantees.",
			`Your tests may work without issue, but we can not guarantee compatibility outside of the above version range.`
		].join("\n");
		log.warn(message);
	}
}
/**
* Ensures that the specified compatibility feature is enabled for Vitest to work.
* @param compatibilityFlags The list of current compatibility flags.
* @param feature The name of the feature to enable.
*/
function ensureFeature(compatibilityFlags, feature) {
	const flagToEnable = `enable_${feature}`;
	const flagToDisable = `disable_${feature}`;
	if (!compatibilityFlags.includes(flagToEnable)) {
		log.debug(`Adding \`${flagToEnable}\` compatibility flag during tests as this feature is needed to support the Vitest runner.`);
		compatibilityFlags.push(flagToEnable);
	}
	if (compatibilityFlags.includes(flagToDisable)) {
		log.info(`Removing \`${flagToDisable}\` compatibility flag during tests as that feature is needed to support the Vitest runner.`);
		compatibilityFlags.splice(compatibilityFlags.indexOf(flagToDisable), 1);
	}
}

//#endregion
export { assertCompatibleVitestVersion, buildPagesASSETSBinding, cloudflarePool, cloudflareTest, connectToMiniflareSocket, getDurableObjectDesignators, getProjectMiniflare, getRunnerName, maybeGetResolvedMainPath, poolWorkerStarted, poolWorkerStopped, readD1Migrations, structuredSerializableParse, structuredSerializableStringify };
